{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "ImportError",
     "evalue": "No module named Bio",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mImportError\u001b[0m                               Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-6-e4eec7e8ae45>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      7\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mos\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      8\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mos\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 9\u001b[1;33m \u001b[1;32mfrom\u001b[0m \u001b[0mBio\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mmotifs\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     10\u001b[0m \u001b[1;32mfrom\u001b[0m \u001b[0mBio\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mSeqIO\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     11\u001b[0m \u001b[1;32mfrom\u001b[0m \u001b[0mBio\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mAlphabet\u001b[0m \u001b[1;32mimport\u001b[0m \u001b[0mIUPAC\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mImportError\u001b[0m: No module named Bio"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import math\n",
    "import matplotlib.pyplot as plt\n",
    "import re\n",
    "import itertools\n",
    "import os\n",
    "import os.path\n",
    "from Bio import motifs\n",
    "from Bio import SeqIO\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio.Seq import Seq\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene</th>\n",
       "      <th>score</th>\n",
       "      <th>nexpt</th>\n",
       "      <th>expts</th>\n",
       "      <th>expscores</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10073</td>\n",
       "      <td>10329</td>\n",
       "      <td>ZBTB33</td>\n",
       "      <td>354</td>\n",
       "      <td>2</td>\n",
       "      <td>204,246</td>\n",
       "      <td>354,138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10149</td>\n",
       "      <td>10413</td>\n",
       "      <td>CEBPB</td>\n",
       "      <td>201</td>\n",
       "      <td>1</td>\n",
       "      <td>343</td>\n",
       "      <td>201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16110</td>\n",
       "      <td>16390</td>\n",
       "      <td>CTCF</td>\n",
       "      <td>227</td>\n",
       "      <td>7</td>\n",
       "      <td>213,612,621,627,628,631,662</td>\n",
       "      <td>110,139,171,209,227,200,170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>29198</td>\n",
       "      <td>29688</td>\n",
       "      <td>TAF1</td>\n",
       "      <td>184</td>\n",
       "      <td>1</td>\n",
       "      <td>157</td>\n",
       "      <td>184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>29275</td>\n",
       "      <td>29591</td>\n",
       "      <td>GABPA</td>\n",
       "      <td>198</td>\n",
       "      <td>1</td>\n",
       "      <td>180</td>\n",
       "      <td>198</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname  start    end    gene  score  nexpt                        expts  \\\n",
       "0    chr1  10073  10329  ZBTB33    354      2                      204,246   \n",
       "1    chr1  10149  10413   CEBPB    201      1                          343   \n",
       "2    chr1  16110  16390    CTCF    227      7  213,612,621,627,628,631,662   \n",
       "3    chr1  29198  29688    TAF1    184      1                          157   \n",
       "4    chr1  29275  29591   GABPA    198      1                          180   \n",
       "\n",
       "                     expscores  \n",
       "0                      354,138  \n",
       "1                          201  \n",
       "2  110,139,171,209,227,200,170  \n",
       "3                          184  \n",
       "4                          198  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get ENCODE TF binding sites\n",
    "tfbs = pd.read_table('c:/ravip/encode/wgEncodeRegTfbsClusteredV3.bed.gz',\n",
    "                     header=None,names=['seqname','start','end','gene','score', 'nexpt', 'expts', 'expscores'],\n",
    "                     dtype={'start':np.int, 'end':np.int})\n",
    "tfbs[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RNAFile</th>\n",
       "      <th>DNAFile</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>Center</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>...</th>\n",
       "      <th>tumorRNAFraction</th>\n",
       "      <th>tumorDNAMultiple</th>\n",
       "      <th>tumorRNAMultiple</th>\n",
       "      <th>tumorDNARatio</th>\n",
       "      <th>tumorRNARatio</th>\n",
       "      <th>FractionOfMeanExpression</th>\n",
       "      <th>zOfmeanExpression</th>\n",
       "      <th>IsSingle</th>\n",
       "      <th>CancerType</th>\n",
       "      <th>gender</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hnsc\\01510fa3-1918-45a8-836e-6a1b009b53e7-CDKN...</td>\n",
       "      <td>hnsc\\0811a03f-37c3-480d-a207-058f2fd784fb-CDKN...</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>1029</td>\n",
       "      <td>broad.mit.edu</td>\n",
       "      <td>37</td>\n",
       "      <td>9</td>\n",
       "      <td>21971120</td>\n",
       "      <td>21971120</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>0.660000</td>\n",
       "      <td>1.727273</td>\n",
       "      <td>1.941176</td>\n",
       "      <td>1.727273</td>\n",
       "      <td>1.941176</td>\n",
       "      <td>0.224637</td>\n",
       "      <td>-0.522650</td>\n",
       "      <td>True</td>\n",
       "      <td>hnsc</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lusc\\01e3b5df-7733-445a-aaa4-ce0809f04736-CDKN...</td>\n",
       "      <td>lusc\\904a0fce-3a41-4716-aa0b-5455f82d8e62-CDKN...</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>1029</td>\n",
       "      <td>broad.mit.edu</td>\n",
       "      <td>37</td>\n",
       "      <td>9</td>\n",
       "      <td>21971017</td>\n",
       "      <td>21971017</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>0.886313</td>\n",
       "      <td>2.083333</td>\n",
       "      <td>7.796053</td>\n",
       "      <td>2.083333</td>\n",
       "      <td>7.796053</td>\n",
       "      <td>2.165528</td>\n",
       "      <td>1.007740</td>\n",
       "      <td>True</td>\n",
       "      <td>lusc</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>coad\\1c3158f8-0ab0-4205-b344-c39b060550ca-CDKN...</td>\n",
       "      <td>coad\\9b09830c-ee71-440e-acd4-c75a3bc76e5a-CDKN...</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>1029</td>\n",
       "      <td>hgsc.bcm.edu</td>\n",
       "      <td>37</td>\n",
       "      <td>9</td>\n",
       "      <td>21971058</td>\n",
       "      <td>21971058</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>0.851852</td>\n",
       "      <td>-1.555556</td>\n",
       "      <td>5.750000</td>\n",
       "      <td>0.642857</td>\n",
       "      <td>5.750000</td>\n",
       "      <td>2.084673</td>\n",
       "      <td>0.772367</td>\n",
       "      <td>True</td>\n",
       "      <td>other</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hnsc\\0a250cb6-7a84-4924-86ee-623793b6abac-CDKN...</td>\n",
       "      <td>hnsc\\72cf828e-8158-43da-adee-f6d2e1b0d876-CDKN...</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>1029</td>\n",
       "      <td>broad.mit.edu</td>\n",
       "      <td>37</td>\n",
       "      <td>9</td>\n",
       "      <td>21971035</td>\n",
       "      <td>21971035</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>0.975845</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>40.400000</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>40.400000</td>\n",
       "      <td>2.162274</td>\n",
       "      <td>0.783455</td>\n",
       "      <td>True</td>\n",
       "      <td>hnsc</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hnsc\\0b6f58df-8922-4ba4-9614-eb2bdf40d2f9-CDKN...</td>\n",
       "      <td>hnsc\\6662c862-34be-48e2-9823-066d2c239cfd-CDKN...</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>1029</td>\n",
       "      <td>broad.mit.edu</td>\n",
       "      <td>37</td>\n",
       "      <td>9</td>\n",
       "      <td>21971029</td>\n",
       "      <td>21971029</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>0.887218</td>\n",
       "      <td>1.958333</td>\n",
       "      <td>7.866667</td>\n",
       "      <td>1.958333</td>\n",
       "      <td>7.866667</td>\n",
       "      <td>0.533013</td>\n",
       "      <td>-0.314783</td>\n",
       "      <td>True</td>\n",
       "      <td>hnsc</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 58 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             RNAFile  \\\n",
       "0  hnsc\\01510fa3-1918-45a8-836e-6a1b009b53e7-CDKN...   \n",
       "1  lusc\\01e3b5df-7733-445a-aaa4-ce0809f04736-CDKN...   \n",
       "2  coad\\1c3158f8-0ab0-4205-b344-c39b060550ca-CDKN...   \n",
       "3  hnsc\\0a250cb6-7a84-4924-86ee-623793b6abac-CDKN...   \n",
       "4  hnsc\\0b6f58df-8922-4ba4-9614-eb2bdf40d2f9-CDKN...   \n",
       "\n",
       "                                             DNAFile Hugo_Symbol  \\\n",
       "0  hnsc\\0811a03f-37c3-480d-a207-058f2fd784fb-CDKN...      CDKN2A   \n",
       "1  lusc\\904a0fce-3a41-4716-aa0b-5455f82d8e62-CDKN...      CDKN2A   \n",
       "2  coad\\9b09830c-ee71-440e-acd4-c75a3bc76e5a-CDKN...      CDKN2A   \n",
       "3  hnsc\\72cf828e-8158-43da-adee-f6d2e1b0d876-CDKN...      CDKN2A   \n",
       "4  hnsc\\6662c862-34be-48e2-9823-066d2c239cfd-CDKN...      CDKN2A   \n",
       "\n",
       "   Entrez_Gene_Id         Center NCBI_Build  Chromosome  Start_Position  \\\n",
       "0            1029  broad.mit.edu         37           9        21971120   \n",
       "1            1029  broad.mit.edu         37           9        21971017   \n",
       "2            1029   hgsc.bcm.edu         37           9        21971058   \n",
       "3            1029  broad.mit.edu         37           9        21971035   \n",
       "4            1029  broad.mit.edu         37           9        21971029   \n",
       "\n",
       "   End_Position Strand   ...   tumorRNAFraction tumorDNAMultiple  \\\n",
       "0      21971120      +   ...           0.660000         1.727273   \n",
       "1      21971017      +   ...           0.886313         2.083333   \n",
       "2      21971058      +   ...           0.851852        -1.555556   \n",
       "3      21971035      +   ...           0.975845         5.000000   \n",
       "4      21971029      +   ...           0.887218         1.958333   \n",
       "\n",
       "  tumorRNAMultiple tumorDNARatio tumorRNARatio FractionOfMeanExpression  \\\n",
       "0         1.941176      1.727273      1.941176                 0.224637   \n",
       "1         7.796053      2.083333      7.796053                 2.165528   \n",
       "2         5.750000      0.642857      5.750000                 2.084673   \n",
       "3        40.400000      5.000000     40.400000                 2.162274   \n",
       "4         7.866667      1.958333      7.866667                 0.533013   \n",
       "\n",
       "  zOfmeanExpression IsSingle CancerType  gender  \n",
       "0         -0.522650     True       hnsc    male  \n",
       "1          1.007740     True       lusc    male  \n",
       "2          0.772367     True      other  female  \n",
       "3          0.783455     True       hnsc  female  \n",
       "4         -0.314783     True       hnsc    male  \n",
       "\n",
       "[5 rows x 58 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cdkn2a = pd.read_table('data/cdkn2a.tsv')\n",
    "cdkn2a[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "p53 = pd.read_table('data/tp53.tsv')\n",
    "p53[:5]\n",
    "muts = p53"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def read_vcf(filename, nrows=None, somatic=False, donor=None, skipcols=None):\n",
    "    names=['chrom','pos','id','ref','alt','qual','filter','info']\n",
    "    if somatic:\n",
    "        names = names + ['format','normal','tumor']\n",
    "    usecols = None if skipcols==None else [i for i in range(len(names)) if names[i] not in skipcols]\n",
    "    alldna = pd.read_table(filename, names=names,\n",
    "                    dtype={'chrom':np.character}, comment='#', index_col=False, nrows=nrows, usecols=usecols)\n",
    "    if not alldna.iloc[0].chrom.startswith('chr'):\n",
    "        alldna.chrom = alldna.chrom.apply(lambda c: 'chr'+c)\n",
    "    if donor != None:\n",
    "        alldna['donor'] = donor\n",
    "    return alldna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>id</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>qual</th>\n",
       "      <th>filter</th>\n",
       "      <th>info</th>\n",
       "      <th>donor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10144</td>\n",
       "      <td>MU29575239</td>\n",
       "      <td>TA</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>CONSEQUENCE=||||||intergenic_region||,WASH7P|E...</td>\n",
       "      <td>ravi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10151</td>\n",
       "      <td>MU15890267</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>CONSEQUENCE=||||||intergenic_region||,WASH7P|E...</td>\n",
       "      <td>ravi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10175</td>\n",
       "      <td>MU23014170</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>CONSEQUENCE=||||||intergenic_region||,WASH7P|E...</td>\n",
       "      <td>ravi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10181</td>\n",
       "      <td>MU4819728</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>CONSEQUENCE=||||||intergenic_region||,WASH7P|E...</td>\n",
       "      <td>ravi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>10237</td>\n",
       "      <td>MU4630481</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>CONSEQUENCE=||||||intergenic_region||,WASH7P|E...</td>\n",
       "      <td>ravi</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom    pos          id ref alt qual filter  \\\n",
       "0  chr1  10144  MU29575239  TA   T    .      .   \n",
       "1  chr1  10151  MU15890267   T   A    .      .   \n",
       "2  chr1  10175  MU23014170   T   A    .      .   \n",
       "3  chr1  10181   MU4819728   A   T    .      .   \n",
       "4  chr1  10237   MU4630481   A   C    .      .   \n",
       "\n",
       "                                                info donor  \n",
       "0  CONSEQUENCE=||||||intergenic_region||,WASH7P|E...  ravi  \n",
       "1  CONSEQUENCE=||||||intergenic_region||,WASH7P|E...  ravi  \n",
       "2  CONSEQUENCE=||||||intergenic_region||,WASH7P|E...  ravi  \n",
       "3  CONSEQUENCE=||||||intergenic_region||,WASH7P|E...  ravi  \n",
       "4  CONSEQUENCE=||||||intergenic_region||,WASH7P|E...  ravi  "
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# not useful - only has frequencies, not mapping to individual donors\n",
    "simple = read_vcf('d:/sequence/pancan/simple_somatic_mutation.aggregated.vcf.gz', nrows=10, donor='ravi')\n",
    "simple[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2081 g:\\pancan2\\vcf\\0009b464-b376-4fbc-8a56-da538269a02f.svcp_1-0-4.20150204.somatic.snv_mnv.vcf.gz g:\\pancan2\\vcf\\fe8b7893-4682-4a85-a520-8e1d654d037a\\1bdfaa8a-1b7e-4251-b135-ce5e4568371b.svcp_1-0-4.20150116.somatic.snv_mnv.vcf.gz\n"
     ]
    }
   ],
   "source": [
    "with open('g:/pancan/somatic.txt', 'r') as f:\n",
    "    vcfs = f.read().split('\\n')\n",
    "vcfs = vcfs[:-1]\n",
    "print len(vcfs), vcfs[0], vcfs[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# get a set of TF binding matrices\n",
    "with open('../regmut/pfm_vertebrates.txt', 'r') as h:\n",
    "    jaspar = {m.name.strip().split(':')[0].upper(): m.counts.normalize().log_odds() for m in motifs.parse(h, 'jaspar')}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "161 196 80\n"
     ]
    }
   ],
   "source": [
    "# reduce tfbs to those for which we have matrices\n",
    "print len(set(tfbs['gene'])), len(set(jaspar.keys())), len(set(tfbs['gene']).intersection(set(jaspar.keys())))\n",
    "common = set(tfbs['gene']).intersection(set(jaspar.keys()))\n",
    "tfbs2 = tfbs[tfbs.gene.isin(common)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def score_mut(m):\n",
    "    mx = jaspar[m['gene']]\n",
    "    n = len(mx[mx.keys()[0]])\n",
    "    p = int(m['pos'])\n",
    "    refseq = hg19[m['chr']][p-n-1:p+n]\n",
    "    refscore = reduce(max, (score for pos, score in mx.search(refseq, threshold=0.5) if score > 0 and pos >=0 and pos < n), 0)\n",
    "    mutseq = refseq[:n] + m['var_allele'] + refseq[n+1:]\n",
    "    mutscore = reduce(max, (score for pos, score in mx.search(mutseq, threshold=0.5) if score > 0 and pos >=0 and pos < n), 0)\n",
    "    return (mutscore - refscore) / mx.max"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# get reference genome\n",
    "hg19 = {contig.name: Seq(str(contig.seq), IUPAC.unambiguous_dna)\n",
    "        for contig in SeqIO.parse('d:/sequence/genomes/hg19.fa', 'fasta')}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# left join by genomic regions\n",
    "# cols[12] are names of chrom, start, end followed by names of columns to include in join result\n",
    "# limitations:\n",
    "# genomic range columns (rcols1, rcols2) must be distinct\n",
    "# only works if region sets are non-overlapping\n",
    "# drops duplicates\n",
    "def merge_regions(df1, cols1, df2, cols2):\n",
    "    result = None\n",
    "    [seqname1, start1, end1] = cols1[:3]\n",
    "    [seqname2, start2, end2] = cols2[:3]\n",
    "    cols1 = list(pd.Series(cols1).drop_duplicates())\n",
    "    cols2 = list(pd.Series(cols2).drop_duplicates())\n",
    "    for contig in set(df1[seqname1]).intersection(set(df2[seqname2])):\n",
    "        c1 = df1[df1[seqname1]==contig].sort(end1).reset_index(drop=True)\n",
    "        c2 = df2[df2[seqname2]==contig].sort(start2).drop_duplicates(start2).reset_index(drop=True)\n",
    "        imerge = pd.cut(c1[end1], c2[start2], labels=False)\n",
    "        keep = imerge >= 0\n",
    "        cmerge = pd.concat([c1[cols1][keep], c2[cols2].iloc[imerge[keep]].reset_index(drop=True)], axis=1).reset_index(drop=True)\n",
    "        cmerge = cmerge[cmerge[start1] <= cmerge[end2]]\n",
    "        result = cmerge if result is None else pd.concat([result, cmerge])\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                start             end  tumorDNARatio  tumorRNARatio  \\\n",
      "count       62.000000       62.000000      62.000000      62.000000   \n",
      "mean   7550435.612903  7550435.774194       2.239220      85.525777   \n",
      "std      29657.851182    29657.965000       2.680856     273.122036   \n",
      "min    7517622.000000  7517622.000000       0.000100       0.000100   \n",
      "25%    7518239.000000  7518239.000000       0.566770       0.639803   \n",
      "50%    7576852.000000  7576852.000000       1.175000       3.785632   \n",
      "75%    7576855.000000  7576855.000000       2.783596      10.854981   \n",
      "max    7579912.000000  7579912.000000      12.200000    1000.000000   \n",
      "\n",
      "       FractionOfMeanExpression  zOfmeanExpression         tfstart  \\\n",
      "count                 62.000000          62.000000       62.000000   \n",
      "mean                   0.680682          -0.524954  7550293.435484   \n",
      "std                    0.625673           0.984256    29560.420910   \n",
      "min                    0.098993          -2.095353  7517536.000000   \n",
      "25%                    0.216637          -1.100952  7518215.250000   \n",
      "50%                    0.462488          -0.832638  7576605.000000   \n",
      "75%                    0.937713          -0.082651  7576605.000000   \n",
      "max                    2.752274           2.325147  7579881.000000   \n",
      "\n",
      "                tfend  \n",
      "count       62.000000  \n",
      "mean   7550713.500000  \n",
      "std      29419.190371  \n",
      "min    7517957.000000  \n",
      "25%    7518620.000000  \n",
      "50%    7576895.000000  \n",
      "75%    7576895.000000  \n",
      "max    7580205.000000  \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>var_allele</th>\n",
       "      <th>tumorDNARatio</th>\n",
       "      <th>tumorRNARatio</th>\n",
       "      <th>FractionOfMeanExpression</th>\n",
       "      <th>zOfmeanExpression</th>\n",
       "      <th>seqname</th>\n",
       "      <th>tfstart</th>\n",
       "      <th>tfend</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7517622</td>\n",
       "      <td>7517622</td>\n",
       "      <td>G</td>\n",
       "      <td>0.857143</td>\n",
       "      <td>0.127660</td>\n",
       "      <td>0.433537</td>\n",
       "      <td>-1.497759</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7517536</td>\n",
       "      <td>7517957</td>\n",
       "      <td>CTCF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7517742</td>\n",
       "      <td>7517742</td>\n",
       "      <td>C</td>\n",
       "      <td>3.426396</td>\n",
       "      <td>11.823529</td>\n",
       "      <td>0.710731</td>\n",
       "      <td>-0.383841</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7517679</td>\n",
       "      <td>7518336</td>\n",
       "      <td>YY1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7517750</td>\n",
       "      <td>7517750</td>\n",
       "      <td>A</td>\n",
       "      <td>4.407407</td>\n",
       "      <td>2.058824</td>\n",
       "      <td>0.152573</td>\n",
       "      <td>-1.124477</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7517679</td>\n",
       "      <td>7518336</td>\n",
       "      <td>YY1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7517783</td>\n",
       "      <td>7517783</td>\n",
       "      <td>C</td>\n",
       "      <td>1.379747</td>\n",
       "      <td>0.761905</td>\n",
       "      <td>0.167333</td>\n",
       "      <td>-1.104892</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7517777</td>\n",
       "      <td>7518033</td>\n",
       "      <td>REST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7517806</td>\n",
       "      <td>7517806</td>\n",
       "      <td>T</td>\n",
       "      <td>8.238095</td>\n",
       "      <td>4.520833</td>\n",
       "      <td>1.778745</td>\n",
       "      <td>2.059043</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7517786</td>\n",
       "      <td>7518620</td>\n",
       "      <td>E2F1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr    start      end var_allele  tumorDNARatio  tumorRNARatio  \\\n",
       "5  chr17  7517622  7517622          G       0.857143       0.127660   \n",
       "6  chr17  7517742  7517742          C       3.426396      11.823529   \n",
       "7  chr17  7517750  7517750          A       4.407407       2.058824   \n",
       "8  chr17  7517783  7517783          C       1.379747       0.761905   \n",
       "9  chr17  7517806  7517806          T       8.238095       4.520833   \n",
       "\n",
       "   FractionOfMeanExpression  zOfmeanExpression seqname  tfstart    tfend  gene  \n",
       "5                  0.433537          -1.497759   chr17  7517536  7517957  CTCF  \n",
       "6                  0.710731          -0.383841   chr17  7517679  7518336   YY1  \n",
       "7                  0.152573          -1.124477   chr17  7517679  7518336   YY1  \n",
       "8                  0.167333          -1.104892   chr17  7517777  7518033  REST  \n",
       "9                  1.778745           2.059043   chr17  7517786  7518620  E2F1  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "muts_tf = merge_regions(muts, ['chr','start','end','var_allele','tumorDNARatio','tumorRNARatio','FractionOfMeanExpression','zOfmeanExpression'],\n",
    "             tfbs2.rename(columns={'start':'tfstart','end':'tfend'}), ['seqname', 'tfstart', 'tfend', 'gene'])\n",
    "print muts_tf.describe()\n",
    "muts_tf[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "muts_tf['tfscore'] = muts_tf.rename(columns={'start':'pos'}).apply(score_mut, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Line_Number</th>\n",
       "      <th>n_DNA_Matching_Reference</th>\n",
       "      <th>n_DNA_Matching_Tumor</th>\n",
       "      <th>n_DNA_Matching_Neither</th>\n",
       "      <th>n_DNA_Matching_Both</th>\n",
       "      <th>n_RNA_Matching_Reference</th>\n",
       "      <th>...</th>\n",
       "      <th>tumorRNAFraction</th>\n",
       "      <th>tumorDNAMultiple</th>\n",
       "      <th>tumorRNAMultiple</th>\n",
       "      <th>tumorDNARatio</th>\n",
       "      <th>tumorRNARatio</th>\n",
       "      <th>FractionOfMeanExpression</th>\n",
       "      <th>zOfmeanExpression</th>\n",
       "      <th>IsSingle</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>2218.000000</td>\n",
       "      <td>2218</td>\n",
       "      <td>2218.000000</td>\n",
       "      <td>2218.000000</td>\n",
       "      <td>2162.000000</td>\n",
       "      <td>2218.000000</td>\n",
       "      <td>2218.000000</td>\n",
       "      <td>2218.000000</td>\n",
       "      <td>2218.000000</td>\n",
       "      <td>2218.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>2218.000000</td>\n",
       "      <td>2218.000000</td>\n",
       "      <td>2218.000000</td>\n",
       "      <td>2218.000000</td>\n",
       "      <td>2218.000000</td>\n",
       "      <td>2215.000000</td>\n",
       "      <td>2215.000000</td>\n",
       "      <td>2218</td>\n",
       "      <td>2218.000000</td>\n",
       "      <td>2218.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>6308.356628</td>\n",
       "      <td>17</td>\n",
       "      <td>7575980.814247</td>\n",
       "      <td>7575981.141118</td>\n",
       "      <td>53562.034228</td>\n",
       "      <td>54.917042</td>\n",
       "      <td>42.519387</td>\n",
       "      <td>7.951758</td>\n",
       "      <td>5.165464</td>\n",
       "      <td>39.779982</td>\n",
       "      <td>...</td>\n",
       "      <td>0.587831</td>\n",
       "      <td>-4.763100</td>\n",
       "      <td>-0.726035</td>\n",
       "      <td>2.447353</td>\n",
       "      <td>14.845367</td>\n",
       "      <td>1.009202</td>\n",
       "      <td>0.005774</td>\n",
       "      <td>0.8498647</td>\n",
       "      <td>7575980.814247</td>\n",
       "      <td>7575981.141118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>2314.294669</td>\n",
       "      <td>0</td>\n",
       "      <td>10218.657726</td>\n",
       "      <td>10218.713277</td>\n",
       "      <td>51383.442592</td>\n",
       "      <td>103.786970</td>\n",
       "      <td>68.526194</td>\n",
       "      <td>29.254640</td>\n",
       "      <td>36.295859</td>\n",
       "      <td>57.547324</td>\n",
       "      <td>...</td>\n",
       "      <td>0.321791</td>\n",
       "      <td>23.235304</td>\n",
       "      <td>22.246515</td>\n",
       "      <td>30.123471</td>\n",
       "      <td>101.097243</td>\n",
       "      <td>0.647716</td>\n",
       "      <td>1.044555</td>\n",
       "      <td>0.3572845</td>\n",
       "      <td>10218.657726</td>\n",
       "      <td>10218.713277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>17</td>\n",
       "      <td>7514707.000000</td>\n",
       "      <td>7514707.000000</td>\n",
       "      <td>32.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-567.000000</td>\n",
       "      <td>-662.000000</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>0.073070</td>\n",
       "      <td>-2.806443</td>\n",
       "      <td>False</td>\n",
       "      <td>7514707.000000</td>\n",
       "      <td>7514707.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>7157.000000</td>\n",
       "      <td>17</td>\n",
       "      <td>7577120.000000</td>\n",
       "      <td>7577120.000000</td>\n",
       "      <td>10781.250000</td>\n",
       "      <td>14.000000</td>\n",
       "      <td>13.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.339961</td>\n",
       "      <td>-3.290033</td>\n",
       "      <td>-1.941518</td>\n",
       "      <td>0.303950</td>\n",
       "      <td>0.515062</td>\n",
       "      <td>0.530168</td>\n",
       "      <td>-0.757474</td>\n",
       "      <td>1</td>\n",
       "      <td>7577120.000000</td>\n",
       "      <td>7577120.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>7157.000000</td>\n",
       "      <td>17</td>\n",
       "      <td>7577574.000000</td>\n",
       "      <td>7577574.000000</td>\n",
       "      <td>36853.000000</td>\n",
       "      <td>29.000000</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>4.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>22.500000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.704791</td>\n",
       "      <td>-1.364510</td>\n",
       "      <td>2.387427</td>\n",
       "      <td>0.732864</td>\n",
       "      <td>2.387427</td>\n",
       "      <td>0.935619</td>\n",
       "      <td>-0.109412</td>\n",
       "      <td>1</td>\n",
       "      <td>7577574.000000</td>\n",
       "      <td>7577574.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>7157.000000</td>\n",
       "      <td>17</td>\n",
       "      <td>7578406.000000</td>\n",
       "      <td>7578406.000000</td>\n",
       "      <td>79950.750000</td>\n",
       "      <td>56.750000</td>\n",
       "      <td>46.000000</td>\n",
       "      <td>8.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>48.750000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.863636</td>\n",
       "      <td>1.691106</td>\n",
       "      <td>6.333333</td>\n",
       "      <td>1.691106</td>\n",
       "      <td>6.333333</td>\n",
       "      <td>1.324185</td>\n",
       "      <td>0.537331</td>\n",
       "      <td>1</td>\n",
       "      <td>7578406.000000</td>\n",
       "      <td>7578406.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>7157.000000</td>\n",
       "      <td>17</td>\n",
       "      <td>7579912.000000</td>\n",
       "      <td>7579912.000000</td>\n",
       "      <td>314493.000000</td>\n",
       "      <td>1917.000000</td>\n",
       "      <td>1188.000000</td>\n",
       "      <td>909.000000</td>\n",
       "      <td>905.000000</td>\n",
       "      <td>894.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>68.000000</td>\n",
       "      <td>136.000000</td>\n",
       "      <td>1000.000000</td>\n",
       "      <td>1000.000000</td>\n",
       "      <td>5.861530</td>\n",
       "      <td>6.189130</td>\n",
       "      <td>True</td>\n",
       "      <td>7579912.000000</td>\n",
       "      <td>7579912.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Entrez_Gene_Id  Chromosome  Start_Position    End_Position  \\\n",
       "count     2218.000000        2218     2218.000000     2218.000000   \n",
       "mean      6308.356628          17  7575980.814247  7575981.141118   \n",
       "std       2314.294669           0    10218.657726    10218.713277   \n",
       "min          0.000000          17  7514707.000000  7514707.000000   \n",
       "25%       7157.000000          17  7577120.000000  7577120.000000   \n",
       "50%       7157.000000          17  7577574.000000  7577574.000000   \n",
       "75%       7157.000000          17  7578406.000000  7578406.000000   \n",
       "max       7157.000000          17  7579912.000000  7579912.000000   \n",
       "\n",
       "         Line_Number  n_DNA_Matching_Reference  n_DNA_Matching_Tumor  \\\n",
       "count    2162.000000               2218.000000           2218.000000   \n",
       "mean    53562.034228                 54.917042             42.519387   \n",
       "std     51383.442592                103.786970             68.526194   \n",
       "min        32.000000                  0.000000              0.000000   \n",
       "25%     10781.250000                 14.000000             13.000000   \n",
       "50%     36853.000000                 29.000000             25.000000   \n",
       "75%     79950.750000                 56.750000             46.000000   \n",
       "max    314493.000000               1917.000000           1188.000000   \n",
       "\n",
       "       n_DNA_Matching_Neither  n_DNA_Matching_Both  n_RNA_Matching_Reference  \\\n",
       "count             2218.000000          2218.000000               2218.000000   \n",
       "mean                 7.951758             5.165464                 39.779982   \n",
       "std                 29.254640            36.295859                 57.547324   \n",
       "min                  0.000000             0.000000                  0.000000   \n",
       "25%                  2.000000             0.000000                 10.000000   \n",
       "50%                  4.000000             0.000000                 22.500000   \n",
       "75%                  8.000000             0.000000                 48.750000   \n",
       "max                909.000000           905.000000                894.000000   \n",
       "\n",
       "            ...        tumorRNAFraction  tumorDNAMultiple  tumorRNAMultiple  \\\n",
       "count       ...             2218.000000       2218.000000       2218.000000   \n",
       "mean        ...                0.587831         -4.763100         -0.726035   \n",
       "std         ...                0.321791         23.235304         22.246515   \n",
       "min         ...                0.000000       -567.000000       -662.000000   \n",
       "25%         ...                0.339961         -3.290033         -1.941518   \n",
       "50%         ...                0.704791         -1.364510          2.387427   \n",
       "75%         ...                0.863636          1.691106          6.333333   \n",
       "max         ...                1.000000         68.000000        136.000000   \n",
       "\n",
       "       tumorDNARatio  tumorRNARatio  FractionOfMeanExpression  \\\n",
       "count    2218.000000    2218.000000               2215.000000   \n",
       "mean        2.447353      14.845367                  1.009202   \n",
       "std        30.123471     101.097243                  0.647716   \n",
       "min         0.000100       0.000100                  0.073070   \n",
       "25%         0.303950       0.515062                  0.530168   \n",
       "50%         0.732864       2.387427                  0.935619   \n",
       "75%         1.691106       6.333333                  1.324185   \n",
       "max      1000.000000    1000.000000                  5.861530   \n",
       "\n",
       "       zOfmeanExpression   IsSingle           start             end  \n",
       "count        2215.000000       2218     2218.000000     2218.000000  \n",
       "mean            0.005774  0.8498647  7575980.814247  7575981.141118  \n",
       "std             1.044555  0.3572845    10218.657726    10218.713277  \n",
       "min            -2.806443      False  7514707.000000  7514707.000000  \n",
       "25%            -0.757474          1  7577120.000000  7577120.000000  \n",
       "50%            -0.109412          1  7577574.000000  7577574.000000  \n",
       "75%             0.537331          1  7578406.000000  7578406.000000  \n",
       "max             6.189130       True  7579912.000000  7579912.000000  \n",
       "\n",
       "[8 rows x 24 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#muts_tf.tfscore.describe()\n",
    "muts.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Project code</th>\n",
       "      <td>BLCA-US</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Submitter donor ID</th>\n",
       "      <td>096b4f32-10c1-4737-a0dd-cae04c54ee33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Data train</th>\n",
       "      <td>train2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Train2 pilot</th>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Normal WGS submitter specimen ID</th>\n",
       "      <td>27461a27-26eb-4c2c-9c54-e16fbd32c615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Normal WGS submitter sample ID</th>\n",
       "      <td>e0fccaf5-925a-41f9-b87c-cd5ee4aecb59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Normal WGS aliquot ID</th>\n",
       "      <td>e0fccaf5-925a-41f9-b87c-cd5ee4aecb59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Normal WGS alignment GNOS repo(s)</th>\n",
       "      <td>https://cghub.ucsc.edu/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Normal WGS alignment GNOS analysis ID</th>\n",
       "      <td>f22a72c5-73c8-478d-b03e-04599b9d5321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Normal WGS alignment BAM file name</th>\n",
       "      <td>PCAWG.67455c36-aa47-4cc4-8b6d-9a9012b616ed.bam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tumour WGS Specimen Count</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tumour WGS submitter specimen ID(s)</th>\n",
       "      <td>52f538ef-b05d-4c76-9976-ce6d49158016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tumour WGS submitter sample ID(s)</th>\n",
       "      <td>301d6ce3-4099-4c1d-8e50-c04b7ce91450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tumour WGS aliquot ID(s)</th>\n",
       "      <td>301d6ce3-4099-4c1d-8e50-c04b7ce91450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tumour WGS alignment GNOS repo(s)</th>\n",
       "      <td>https://cghub.ucsc.edu/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tumour WGS alignment GNOS analysis ID(s)</th>\n",
       "      <td>a6e8dd23-c8a5-445a-ae4b-b9f92ed6a73e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tumour WGS alignment BAM file name(s)</th>\n",
       "      <td>PCAWG.a833602e-7ee7-442b-9bf9-f6489cd64ba2.bam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sanger variant call GNOS repo(s)</th>\n",
       "      <td>https://gtrepo-osdc-tcga.annailabs.com/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sanger variant call GNOS analysis ID(s)</th>\n",
       "      <td>5230346f-987e-44c3-b9fe-dddb5829e247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sanger variant call GNOS file name prefix</th>\n",
       "      <td>301d6ce3-4099-4c1d-8e50-c04b7ce91450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Normal RNA-Seq submitter specimen ID</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Normal RNA-Seq submitter sample ID</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Normal RNA-Seq aliquot ID</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Normal RNA-Seq STAR alignment GNOS repo(s)</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Normal RNA-Seq STAR alignment GNOS analysis ID</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Normal RNA-Seq STAR alignment BAM file name</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Normal RNA-Seq TopHat2 alignment GNOS repo(s)</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Normal RNA-Seq TopHat2 alignment GNOS analysis ID</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Normal RNA-Seq TopHat2 alignment BAM file name</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tumour RNA-Seq submitter specimen ID</th>\n",
       "      <td>52f538ef-b05d-4c76-9976-ce6d49158016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tumour RNA-Seq submitter sample ID</th>\n",
       "      <td>22e154de-0e3b-443b-8420-48d68d6c1ce4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tumour RNA-Seq aliquot ID</th>\n",
       "      <td>22e154de-0e3b-443b-8420-48d68d6c1ce4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tumour RNA-Seq STAR alignment GNOS repo(s)</th>\n",
       "      <td>https://cghub.ucsc.edu/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tumour RNA-Seq STAR alignment GNOS analysis ID</th>\n",
       "      <td>a63c9c0a-c8aa-44dc-8915-08f55109a5c9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tumour RNA-Seq STAR alignment BAM file name</th>\n",
       "      <td>PCAWG.f4288632-a94f-4777-a5d6-1a76b12abcc5.STA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tumour RNA-Seq TopHat2 alignment GNOS repo(s)</th>\n",
       "      <td>https://cghub.ucsc.edu/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tumour RNA-Seq TopHat2 alignment GNOS analysis ID</th>\n",
       "      <td>81438f7c-ee87-44a7-acd1-562ec8f386f0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tumour RNA-Seq TopHat2 alignment BAM file name</th>\n",
       "      <td>PCAWG.f4288632-a94f-4777-a5d6-1a76b12abcc5.Top...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                   0\n",
       "Project code                                                                                 BLCA-US\n",
       "Submitter donor ID                                              096b4f32-10c1-4737-a0dd-cae04c54ee33\n",
       "Data train                                                                                    train2\n",
       "Train2 pilot                                                                                    True\n",
       "Normal WGS submitter specimen ID                                27461a27-26eb-4c2c-9c54-e16fbd32c615\n",
       "Normal WGS submitter sample ID                                  e0fccaf5-925a-41f9-b87c-cd5ee4aecb59\n",
       "Normal WGS aliquot ID                                           e0fccaf5-925a-41f9-b87c-cd5ee4aecb59\n",
       "Normal WGS alignment GNOS repo(s)                                            https://cghub.ucsc.edu/\n",
       "Normal WGS alignment GNOS analysis ID                           f22a72c5-73c8-478d-b03e-04599b9d5321\n",
       "Normal WGS alignment BAM file name                    PCAWG.67455c36-aa47-4cc4-8b6d-9a9012b616ed.bam\n",
       "Tumour WGS Specimen Count                                                                          1\n",
       "Tumour WGS submitter specimen ID(s)                             52f538ef-b05d-4c76-9976-ce6d49158016\n",
       "Tumour WGS submitter sample ID(s)                               301d6ce3-4099-4c1d-8e50-c04b7ce91450\n",
       "Tumour WGS aliquot ID(s)                                        301d6ce3-4099-4c1d-8e50-c04b7ce91450\n",
       "Tumour WGS alignment GNOS repo(s)                                            https://cghub.ucsc.edu/\n",
       "Tumour WGS alignment GNOS analysis ID(s)                        a6e8dd23-c8a5-445a-ae4b-b9f92ed6a73e\n",
       "Tumour WGS alignment BAM file name(s)                 PCAWG.a833602e-7ee7-442b-9bf9-f6489cd64ba2.bam\n",
       "Sanger variant call GNOS repo(s)                             https://gtrepo-osdc-tcga.annailabs.com/\n",
       "Sanger variant call GNOS analysis ID(s)                         5230346f-987e-44c3-b9fe-dddb5829e247\n",
       "Sanger variant call GNOS file name prefix                       301d6ce3-4099-4c1d-8e50-c04b7ce91450\n",
       "Normal RNA-Seq submitter specimen ID                                                             NaN\n",
       "Normal RNA-Seq submitter sample ID                                                               NaN\n",
       "Normal RNA-Seq aliquot ID                                                                        NaN\n",
       "Normal RNA-Seq STAR alignment GNOS repo(s)                                                       NaN\n",
       "Normal RNA-Seq STAR alignment GNOS analysis ID                                                   NaN\n",
       "Normal RNA-Seq STAR alignment BAM file name                                                      NaN\n",
       "Normal RNA-Seq TopHat2 alignment GNOS repo(s)                                                    NaN\n",
       "Normal RNA-Seq TopHat2 alignment GNOS analysis ID                                                NaN\n",
       "Normal RNA-Seq TopHat2 alignment BAM file name                                                   NaN\n",
       "Tumour RNA-Seq submitter specimen ID                            52f538ef-b05d-4c76-9976-ce6d49158016\n",
       "Tumour RNA-Seq submitter sample ID                              22e154de-0e3b-443b-8420-48d68d6c1ce4\n",
       "Tumour RNA-Seq aliquot ID                                       22e154de-0e3b-443b-8420-48d68d6c1ce4\n",
       "Tumour RNA-Seq STAR alignment GNOS repo(s)                                   https://cghub.ucsc.edu/\n",
       "Tumour RNA-Seq STAR alignment GNOS analysis ID                  a63c9c0a-c8aa-44dc-8915-08f55109a5c9\n",
       "Tumour RNA-Seq STAR alignment BAM file name        PCAWG.f4288632-a94f-4777-a5d6-1a76b12abcc5.STA...\n",
       "Tumour RNA-Seq TopHat2 alignment GNOS repo(s)                                https://cghub.ucsc.edu/\n",
       "Tumour RNA-Seq TopHat2 alignment GNOS analysis ID               81438f7c-ee87-44a7-acd1-562ec8f386f0\n",
       "Tumour RNA-Seq TopHat2 alignment BAM file name     PCAWG.f4288632-a94f-4777-a5d6-1a76b12abcc5.Top..."
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scfiles = pd.read_table('d:/sequence/pancan/santa_cruz_pilot.v2.2015_0504.tsv')\n",
    "scfiles[:1].transpose()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'g:/pancan2/vcf/00db4dc2-3ec7-4ff9-9233-d69c8c8a607f.svcp_1-0-5.20150403.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/020fab36-c7de-4933-b2bf-dc7b019a1326.svcp_1-0-5.20150324.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/0332b017-17d5-4083-8fc4-9d6f8fdbbbde.svcp_1-0-5.20150308.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/03ced0ce-186a-4349-8d98-572c2bc90382.svcp_1-0-5.20150401.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/04339769-517c-448d-a7ca-951f83608c60.svcp_1-0-5.20150305.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/0448206f-3ade-4087-b1a9-4fb2d14e1367.svcp_1-0-5.20150406.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/04aa6b77-8074-480c-872e-a1a47afa5314.svcp_1-0-5.20150324.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/0980e7fd-051d-45e9-9ca6-2baf073da4e8.svcp_1-0-5.20150306.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/09bc8542-4ea1-4599-91b3-d606425b37e2.svcp_1-0-5.20150328.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/0bbc7ede-5022-4084-925c-d65baaf7abc2.svcp_1-0-5.20150307.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/0c39c03a-ec84-42c6-be69-8e98bc7a9f5a.svcp_1-0-5.20150403.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/0c7aca3f-e006-4de3-afc2-20b4f727d4fd.svcp_1-0-5.20150404.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/0dd0718d-5ddf-4c59-8c47-0f51303daeb5.svcp_1-0-5.20150328.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/1021b60d-f7b2-43b0-b2cc-f282d619d533.svcp_1-0-5.20150317.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/14c5b81d-da49-4db1-9834-77711c2b1d38.svcp_1-0-5.20150403.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/153a64ee-6322-4f2e-b3df-77ef7d3a4cf0.svcp_1-0-5.20150307.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/15e097e3-e468-466c-8e57-b59fba90bbf2.svcp_1-0-5.20150325.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/15f90ef0-831b-40a3-98bd-ec226a9e8b26.svcp_1-0-5.20150405.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/16d33e09-2e21-4da2-8e57-e78ce28c4408.svcp_1-0-5.20150401.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/19406c86-dd55-420c-b6d4-b5ca9d3bfed6.svcp_1-0-5.20150307.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/1c0e384f-7254-4afe-93c0-b3fc6c6a7894.svcp_1-0-5.20150406.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/1e8ae8dd-74c1-44d5-a74e-bfdce801c736.svcp_1-0-5.20150401.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/2045c788-9ea8-4ea5-a5e3-65fc16a62adb.svcp_1-0-5.20150317.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/24f887e1-ce82-40f3-9674-11102bd076c0.svcp_1-0-5.20150401.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/2b142863-b963-4cc9-8f8f-c72503c93390.svcp_1-0-4.20150309.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/2bff30d5-be79-4686-8164-7a7d9619d3c0.svcp_1-0-5.20150327.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/2deb8f9a-9e0a-46d1-bcb5-902db3b5c8ac.svcp_1-0-5.20150323.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/322f0b01-2118-4dbe-aba1-3875a54ee71b.svcp_1-0-5.20150318.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/36d1a85e-a09b-4537-86e0-eaf1eb03aed8.svcp_1-0-4.20150203.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/387c6519-6529-4074-a5ab-00f8052a5732.svcp_1-0-5.20150313.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/418a3dec-96ff-4719-becb-e1a8260cce2f.svcp_1-0-5.20150404.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/418e916b-7a4e-4fab-8616-15dcec4d79f8.svcp_1-0-5.20150405.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/448af8b4-e071-48b0-a65b-b4ad17afdc0c.svcp_1-0-5.20150307.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/448fe471-3f4e-4dc8-a4e0-6f147dc93abe.svcp_1-0-5.20150404.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/45a7949d-e63f-4956-866c-df51257032de.svcp_1-0-5.20150404.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/460f7427-ebcf-404b-9e9e-e0278d0bca95.svcp_1-0-5.20150407.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/4624fd20-e2ad-4826-b606-a694b04e1076.svcp_1-0-5.20150328.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/47b5d831-5287-4f62-b17a-6e5eff2e4184.svcp_1-0-4.20150212.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/4838b5a9-968c-4178-bffb-3fafe1f6dc09.svcp_1-0-2.20141229.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/4c2ad4a0-5d57-4e27-9f35-058b2f205f50.svcp_1-0-5.20150317.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/4c3840df-9824-40db-879e-6d24adc8c155.svcp_1-0-5.20150317.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/4d4eff82-ede0-47f1-b49a-ada025952114.svcp_1-0-5.20150414.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/4d7f2c74-862b-4aad-98e1-fa831f14a905.svcp_1-0-5.20150325.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/50ee360c-e3df-4888-9999-ebb88e271d08.svcp_1-0-5.20150327.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/522dc834-8367-4f49-b93b-fcba91a2de7e.svcp_1-0-5.20150401.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/52910a60-bb15-4ba5-9d09-50d8ee6a445b.svcp_1-0-5.20150328.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/5310f698-d31a-4628-84b4-dc9ac29722a5.svcp_1-0-5.20150324.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/54a77010-2f16-4da9-8ca0-746a51801a00.svcp_1-0-5.20150324.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/554b9439-7e11-43ec-8bc3-448c59100b9e.svcp_1-0-5.20150403.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/576a1741-8c65-40da-9f0f-b2ca805d70f3.svcp_1-0-5.20150308.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/5dbf3203-ce73-41e4-bf9a-32fc856f73f5.svcp_1-0-5.20150405.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/638e80c7-9a6e-4a32-a621-fc4168e72343.svcp_1-0-5.20150407.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/65723119-bdfe-46f0-b629-c171023abd71.svcp_1-0-5.20150308.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/65f9a820-f026-4a10-8bb7-1fbc3b38321e.svcp_1-0-5.20150403.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/6a7cdaf6-211e-4eb8-a70e-1153d69c45e4.svcp_1-0-5.20150309.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/74039acd-5aca-4c65-818c-3b577d295be0.svcp_1-0-5.20150405.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/75ba6722-1148-4a52-a9ed-68d890238205.svcp_1-0-5.20150325.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/7692e684-4af2-4892-80ad-1f417b6b38fa.svcp_1-0-5.20150328.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/79eb7bba-f0d8-462c-add7-20a2fb7843e1.svcp_1-0-5.20150317.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/7cae6c0b-36fe-411b-bbba-093a4c846d84.svcp_1-0-5.20150306.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/7cdbe0e8-f614-4f54-b864-fd6b39e8ef1c.svcp_1-0-4.20150316.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/7dc5f8ba-0080-43d3-8426-bd527a970761.svcp_1-0-5.20150308.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/7fc7b316-f3d3-4e71-9417-423fe334e6bc.svcp_1-0-5.20150413.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/81a8b064-e735-455f-b2db-af7ae11daac4.svcp_1-0-2.20141229.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/84c77098-03d0-4b22-afb1-797703e85c6c.svcp_1-0-4.20150203.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/8853cbee-7931-49a6-b063-a806943a10ad.svcp_1-0-2.20141230.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/8dd14f0e-8601-4aa1-864c-3c49e768cdd1.svcp_1-0-5.20150403.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/90611290-e44c-438e-91aa-f3b83dae3b71.svcp_1-0-5.20150413.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/935b113e-f5ed-4a07-8e1d-1603daba7f40.svcp_1-0-2.20150105.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/93b51c61-6eea-4228-a102-840a2e118522.svcp_1-0-5.20150313.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/9536f736-63bc-4099-bd54-740f5910f4a8.svcp_1-0-5.20150308.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/987528ac-437a-4eb8-a335-4f2076d5c006.svcp_1-0-5.20150405.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/992d9566-b1da-421a-bf36-d23382b115fa.svcp_1-0-5.20150328.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/9a90f951-ad96-4883-b97f-ac43395dae5d.svcp_1-0-5.20150406.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/9d29543e-8601-4fd0-8e76-3df3de465cab.svcp_1-0-2.20141222.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/a3fd2ee0-3dd0-4992-9a9f-75e487b43441.svcp_1-0-5.20150318.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/abc94013-71f5-4ac6-88a4-01b4ef9f9d2f.svcp_1-0-5.20150307.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/b07bad52-d44c-4b27-900a-960985bfadec.svcp_1-0-5.20150328.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/b27d75ba-5989-4200-bfe9-f1b7d7cf8008.svcp_1-0-4.20150323.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/b2d17671-d2e1-4c97-8b01-a976d5abe1d6.svcp_1-0-5.20150306.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/b60392fb-43d9-4c9c-b91b-ded40492e61c.svcp_1-0-5.20150309.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/b710996c-8d35-4a2c-bc51-389b3e35b0f3.svcp_1-0-5.20150328.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/b73523d7-f5a5-4140-8537-4df4d1ecf465.svcp_1-0-5.20150404.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/bd3e88b3-b37c-4641-85fa-d8125ba324ca.svcp_1-0-5.20150324.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/bf7462a2-394f-4838-bcb6-4d0126fa48b1.svcp_1-0-5.20150328.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/c016d025-3c92-41c5-b846-493b1fcce79e.svcp_1-0-5.20150413.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/c065761d-f775-457f-bda0-4c7c257a701e.svcp_1-0-5.20150325.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/c612570a-881d-421f-98b9-0e4563133744.svcp_1-0-2.20150111.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/c67409b2-ac25-42a0-8543-4636ef132fe4.svcp_1-0-5.20150405.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/cda1a403-16b6-487c-a82a-c377d1d0f89d.svcp_1-0-5.20150404.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/d02597f8-3ac7-4165-a65f-0e134e5d215b.svcp_1-0-5.20150413.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/d252f328-4583-4e97-9a71-bb2885f06f73.svcp_1-0-5.20150326.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/d53c2680-89ac-4f87-a7ba-f4c669546b76.svcp_1-0-5.20150331.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/d8fbb398-d1da-4444-984a-22c8523625da.svcp_1-0-5.20150404.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/dc537fcf-d910-4c4b-8af9-e7da429f2633.svcp_1-0-5.20150309.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/e2774050-1257-4c29-b201-d79ce70fe17e.svcp_1-0-5.20150325.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/e7ebc6fb-0926-4c8a-a67b-0c6b9c1ffaba.svcp_1-0-5.20150309.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/e9032df3-8692-4146-a867-b9b64c9b310a.svcp_1-0-5.20150318.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/ea1cac20-88c1-4257-9cdb-d2890eb2e123.svcp_1-0-5.20150322.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/ea2055d1-7b8b-4aaf-9a70-8802f4deef9b.svcp_1-0-5.20150309.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/ea43434b-197e-48ac-ae2e-46bc7f3776de.svcp_1-0-4.20150130.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/ea6efcd1-11de-45f0-8bde-f1c06ad27e79.svcp_1-0-5.20150407.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/eb59b79f-bd36-4bf4-83e3-2af73bd1d30a.svcp_1-0-5.20150313.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/ed32c725-08ae-48eb-8fa2-719b9aeb7550.svcp_1-0-5.20150407.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/edc8839f-fa6b-41e0-ad49-f688073730bd.svcp_1-0-5.20150307.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/ee6cc68e-8d2a-41ee-82c6-0fecdf7e6259.svcp_1-0-5.20150313.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/f389176f-d8f3-45c2-aae4-7378a3d6fc7f.svcp_1-0-5.20150404.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/f5a97315-1906-4774-980e-0879c6ad368e.svcp_1-0-5.20150325.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/f7dcc2e4-1fc4-4b39-bc6a-720e66116d68.svcp_1-0-5.20150403.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/fc249113-83d4-4abe-8c80-a4f7305dcd91.svcp_1-0-5.20150328.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/fc6d77a9-121b-48ab-a899-713c3d1319a2.svcp_1-0-4.20150316.somatic.snv_mnv.vcf.gz\\ng:/pancan2/vcf/ffb4f42b-58e9-40c3-8963-11804f041375.svcp_1-0-5.20150311.somatic.snv_mnv.vcf.gz'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutsamples = set(muts.Tumor_Sample_UUID)\n",
    "mutscfiles = scfiles[scfiles['Sanger variant call GNOS file name prefix'].apply(lambda f: f in mutsamples)]\n",
    "vcfdir = 'g:/pancan2/vcf'\n",
    "mutscfiles = [vcfdir + '/' + f for f in os.listdir(vcfdir)\n",
    "              if f.split('.')[0] in set(mutscfiles['Sanger variant call GNOS file name prefix'])]\n",
    "len(mutscfiles)\n",
    "'\\n'.join(mutscfiles)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from ipyparallel import Client\n",
    "c = Client()\n",
    "dview = c[:]\n",
    "%px import pandas as pd\n",
    "%px import numpy as np\n",
    "\n",
    "dview.scatter('vcfs', mutscfiles)\n",
    "dview.push(dict(read_vcf=read_vcf))\n",
    "%px variants = [read_vcf(f, somatic=True, donor=f.split('/')[-1].split('.')[0], skipcols=['info', 'format']) for f in vcfs]\n",
    "variants = dview.gather('variants').get()\n",
    "variants = pd.concat(variants)\n",
    "# clean up worker memory\n",
    "%px variants=None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "muts=p53\n",
    "mutvars = muts.rename(columns={'Tumor_Sample_UUID':'donor'}).merge(variants, on='donor', how='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18178112 112\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RNAFile</th>\n",
       "      <td>kich\\e194e19e-971d-4f34-98d9-1c3c4301ecba-TP53...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNAFile</th>\n",
       "      <td>kich\\14c3ab39-e883-404f-a83c-46733e4a4adc-TP53...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <td>TP53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <td>7157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Center</th>\n",
       "      <td>broad.mit.edu;hgsc.bcm.edu;bcgsc.ca</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NCBI_Build</th>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Chromosome</th>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Start_Position</th>\n",
       "      <td>7578208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>End_Position</th>\n",
       "      <td>7578208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Strand</th>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Variant_Classification</th>\n",
       "      <td>Missense_Mutation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Variant_Type</th>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Reference_Allele</th>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tumor_Seq_Allele_1</th>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tumor_Seq_Allele_2</th>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dbSNP_RS</th>\n",
       "      <td>novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dbSNP_Val_Status</th>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <td>TCGA-KO-8416-01A-11D-2310-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Matched_Norm_Sample_Barcode</th>\n",
       "      <td>TCGA-KO-8416-11A-01D-2311-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Match_Norm_Seq_Allele1</th>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Match_Norm_Seq_Allele2</th>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tumor_Validation_Allele1</th>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tumor_Validation_Allele2</th>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Match_Norm_Validation_Allele1</th>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Match_Norm_Validation_Allele2</th>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Verification_Status</th>\n",
       "      <td>Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Validation_Status</th>\n",
       "      <td>Untested</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mutation_Status</th>\n",
       "      <td>Somatic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sequencing_Phase</th>\n",
       "      <td>Phase_I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sequence_Source</th>\n",
       "      <td>WXS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>n_DNA_Matching_Both</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>n_RNA_Matching_Reference</th>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>n_RNA_Matching_Tumor</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>n_RNA_Matching_Neither</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>n_RNA_Matching_Both</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tumorDNAFraction</th>\n",
       "      <td>0.01612903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tumorRNAFraction</th>\n",
       "      <td>0.06666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tumorDNAMultiple</th>\n",
       "      <td>-61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tumorRNAMultiple</th>\n",
       "      <td>-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tumorDNARatio</th>\n",
       "      <td>0.01639344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tumorRNARatio</th>\n",
       "      <td>0.07142857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FractionOfMeanExpression</th>\n",
       "      <td>0.704269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zOfmeanExpression</th>\n",
       "      <td>-0.7382912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IsSingle</th>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CancerType</th>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gender</th>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <td>TP53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>start</th>\n",
       "      <td>7578208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>end</th>\n",
       "      <td>7578208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr</th>\n",
       "      <td>chr17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_allele</th>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrom</th>\n",
       "      <td>chr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pos</th>\n",
       "      <td>10443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <td>149755a0-8ea8-11e4-9350-02a2af1d574b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ref</th>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>alt</th>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>qual</th>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>filter</th>\n",
       "      <td>SR;MN;MQ;VUM;MNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>normal</th>\n",
       "      <td>0|0:0:6:0:1:0:65:1:2:4.0e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tumor</th>\n",
       "      <td>0|1:0:19:0:1:2:118:0:21:1.4e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>72 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                               0\n",
       "RNAFile                        kich\\e194e19e-971d-4f34-98d9-1c3c4301ecba-TP53...\n",
       "DNAFile                        kich\\14c3ab39-e883-404f-a83c-46733e4a4adc-TP53...\n",
       "Hugo_Symbol                                                                 TP53\n",
       "Entrez_Gene_Id                                                              7157\n",
       "Center                                       broad.mit.edu;hgsc.bcm.edu;bcgsc.ca\n",
       "NCBI_Build                                                                    37\n",
       "Chromosome                                                                    17\n",
       "Start_Position                                                           7578208\n",
       "End_Position                                                             7578208\n",
       "Strand                                                                         +\n",
       "Variant_Classification                                         Missense_Mutation\n",
       "Variant_Type                                                                 SNP\n",
       "Reference_Allele                                                               T\n",
       "Tumor_Seq_Allele_1                                                             T\n",
       "Tumor_Seq_Allele_2                                                             A\n",
       "dbSNP_RS                                                                   novel\n",
       "dbSNP_Val_Status                                                             NaN\n",
       "Tumor_Sample_Barcode                                TCGA-KO-8416-01A-11D-2310-10\n",
       "Matched_Norm_Sample_Barcode                         TCGA-KO-8416-11A-01D-2311-10\n",
       "Match_Norm_Seq_Allele1                                                         T\n",
       "Match_Norm_Seq_Allele2                                                         T\n",
       "Tumor_Validation_Allele1                                                       .\n",
       "Tumor_Validation_Allele2                                                       .\n",
       "Match_Norm_Validation_Allele1                                                  .\n",
       "Match_Norm_Validation_Allele2                                                  .\n",
       "Verification_Status                                                      Unknown\n",
       "Validation_Status                                                       Untested\n",
       "Mutation_Status                                                          Somatic\n",
       "Sequencing_Phase                                                         Phase_I\n",
       "Sequence_Source                                                              WXS\n",
       "...                                                                          ...\n",
       "n_DNA_Matching_Both                                                            0\n",
       "n_RNA_Matching_Reference                                                      13\n",
       "n_RNA_Matching_Tumor                                                           1\n",
       "n_RNA_Matching_Neither                                                         1\n",
       "n_RNA_Matching_Both                                                            0\n",
       "tumorDNAFraction                                                      0.01612903\n",
       "tumorRNAFraction                                                      0.06666667\n",
       "tumorDNAMultiple                                                             -61\n",
       "tumorRNAMultiple                                                             -14\n",
       "tumorDNARatio                                                         0.01639344\n",
       "tumorRNARatio                                                         0.07142857\n",
       "FractionOfMeanExpression                                                0.704269\n",
       "zOfmeanExpression                                                     -0.7382912\n",
       "IsSingle                                                                   False\n",
       "CancerType                                                                 other\n",
       "gender                                                                      male\n",
       "gene                                                                        TP53\n",
       "start                                                                    7578208\n",
       "end                                                                      7578208\n",
       "chr                                                                        chr17\n",
       "var_allele                                                                     T\n",
       "chrom                                                                       chr1\n",
       "pos                                                                        10443\n",
       "id                                          149755a0-8ea8-11e4-9350-02a2af1d574b\n",
       "ref                                                                            C\n",
       "alt                                                                            T\n",
       "qual                                                                           .\n",
       "filter                                                          SR;MN;MQ;VUM;MNP\n",
       "normal                                              0|0:0:6:0:1:0:65:1:2:4.0e-02\n",
       "tumor                                            0|1:0:19:0:1:2:118:0:21:1.4e-01\n",
       "\n",
       "[72 rows x 1 columns]"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print len(mutvars.donor), len(set(mutvars.donor))\n",
    "mutvars[:1].transpose()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>enhchr</th>\n",
       "      <th>enhstart</th>\n",
       "      <th>enhend</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>CellLine</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2155200</td>\n",
       "      <td>2155400</td>\n",
       "      <td>NM_003036</td>\n",
       "      <td>GM12878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2155400</td>\n",
       "      <td>2155600</td>\n",
       "      <td>NM_003036</td>\n",
       "      <td>GM12878</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  enhchr  enhstart   enhend  gene_name CellLine\n",
       "0   chr1   2155200  2155400  NM_003036  GM12878\n",
       "1   chr1   2155400  2155600  NM_003036  GM12878"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get enhancer-gene association from Ernst et al\n",
    "def read_ernst(f):\n",
    "    df = pd.read_csv(os.path.join('../regmut/data/ernst', f), names=['enhchr','enhstart','enhend','gene_name'], sep='\\t')\n",
    "    df['CellLine'] = f.split('_')[1]\n",
    "    return df\n",
    "ernstgene = pd.concat((read_ernst(f) for f in os.listdir('../regmut/data/ernst') if f.startswith('links_') & f.endswith('.txt')))\n",
    "promgene = ernstgene\n",
    "promgene[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9 set(['NM_000546'])\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>enhchr</th>\n",
       "      <th>enhstart</th>\n",
       "      <th>enhend</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>CellLine</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5866</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7520200</td>\n",
       "      <td>7520400</td>\n",
       "      <td>NM_000546</td>\n",
       "      <td>Huvec</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5868</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7520400</td>\n",
       "      <td>7520600</td>\n",
       "      <td>NM_000546</td>\n",
       "      <td>Huvec</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5870</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7520600</td>\n",
       "      <td>7520800</td>\n",
       "      <td>NM_000546</td>\n",
       "      <td>Huvec</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5872</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7520800</td>\n",
       "      <td>7521000</td>\n",
       "      <td>NM_000546</td>\n",
       "      <td>Huvec</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5874</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7521000</td>\n",
       "      <td>7521200</td>\n",
       "      <td>NM_000546</td>\n",
       "      <td>Huvec</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5876</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7521200</td>\n",
       "      <td>7521400</td>\n",
       "      <td>NM_000546</td>\n",
       "      <td>Huvec</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5878</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7521400</td>\n",
       "      <td>7521600</td>\n",
       "      <td>NM_000546</td>\n",
       "      <td>Huvec</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5855</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7521600</td>\n",
       "      <td>7521800</td>\n",
       "      <td>NM_000546</td>\n",
       "      <td>HSMM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     enhchr  enhstart   enhend  gene_name CellLine\n",
       "5866  chr17   7520200  7520400  NM_000546    Huvec\n",
       "5868  chr17   7520400  7520600  NM_000546    Huvec\n",
       "5870  chr17   7520600  7520800  NM_000546    Huvec\n",
       "5872  chr17   7520800  7521000  NM_000546    Huvec\n",
       "5874  chr17   7521000  7521200  NM_000546    Huvec\n",
       "5876  chr17   7521200  7521400  NM_000546    Huvec\n",
       "5878  chr17   7521400  7521600  NM_000546    Huvec\n",
       "5855  chr17   7521600  7521800  NM_000546     HSMM"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut_prom = promgene[promgene.gene_name.isin(['NM_000546'])]\n",
    "print len(mut_prom.gene_name), set(mut_prom.gene_name)\n",
    "mut_prom.sort(['enhchr','enhstart','CellLine']).drop_duplicates(['enhchr','enhstart'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#mutvars['chr'] = mutvars['Chromosome'].apply(lambda c: 'chr' + str(c))\n",
    "mut_prom_vars = merge_regions(mut_prom.sort(['enhchr','enhstart']), ['enhchr', 'enhstart', 'enhend', 'CellLine', 'gene_name'],\n",
    "                              mutvars.sort(['chrom','pos']), ['chrom', 'pos', 'pos', 'donor', 'Tumor_Seq_Allele_1',\n",
    "                                        'Tumor_Seq_Allele_2', 'Variant_Classification', 'Variant_Type',\n",
    "                                        'tumorDNARatio', 'tumorRNARatio', 'FractionOfMeanExpression', 'zOfmeanExpression'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def merge_regions(df1, cols1, df2, cols2):\n",
    "    result = None\n",
    "    [seqname1, start1, end1] = cols1[:3]\n",
    "    [seqname2, start2, end2] = cols2[:3]\n",
    "    cols1 = list(pd.Series(cols1).drop_duplicates())\n",
    "    cols2 = list(pd.Series(cols2).drop_duplicates())\n",
    "    df1 = df1[cols1].sort([seqname1, start1, end1])\n",
    "    df2 = df2[cols2].sort([seqname2, start2, end2])\n",
    "    for contig in set(df1[seqname1]).intersection(set(df2[seqname2])):\n",
    "        c1 = df1[df1[seqname1]==contig].sort(end1).reset_index(drop=True)\n",
    "        c2 = df2[df2[seqname2]==contig].sort(start2).drop_duplicates(start2).reset_index(drop=True)\n",
    "        imerge = pd.cut(c1[end1], c2[start2], labels=False)\n",
    "        keep = imerge >= 0\n",
    "        cmerge = pd.concat([c1[cols1][keep], c2[cols2].iloc[imerge[keep]].reset_index(drop=True)], axis=1).reset_index(drop=True)\n",
    "        cmerge = cmerge[cmerge[start1] <= cmerge[end2]]\n",
    "        result = cmerge if result is None else pd.concat([result, cmerge])\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "mvp53 = mutvars[(mutvars.chrom=='chr17') & (mutvars.pos >= 7519000) & (mutvars.pos <= 7524000)]\n",
    "promp53 = mut_prom[(mut_prom.enhchr=='chr17') & (mut_prom.enhstart >= 7519000) & (mut_prom.enhend <= 7524000)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>enhchr</th>\n",
       "      <th>enhstart</th>\n",
       "      <th>enhend</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7520200</td>\n",
       "      <td>7520400</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7519825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7520400</td>\n",
       "      <td>7520600</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7519825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7520600</td>\n",
       "      <td>7520800</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7519825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7520800</td>\n",
       "      <td>7521000</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7519825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7521000</td>\n",
       "      <td>7521200</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7521110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7521200</td>\n",
       "      <td>7521400</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7521110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7521400</td>\n",
       "      <td>7521600</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7521110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7521600</td>\n",
       "      <td>7521800</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7521788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7521600</td>\n",
       "      <td>7521800</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7521788</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  enhchr  enhstart   enhend  chrom      pos\n",
       "0  chr17   7520200  7520400  chr17  7519825\n",
       "1  chr17   7520400  7520600  chr17  7519825\n",
       "2  chr17   7520600  7520800  chr17  7519825\n",
       "3  chr17   7520800  7521000  chr17  7519825\n",
       "4  chr17   7521000  7521200  chr17  7521110\n",
       "5  chr17   7521200  7521400  chr17  7521110\n",
       "6  chr17   7521400  7521600  chr17  7521110\n",
       "7  chr17   7521600  7521800  chr17  7521788\n",
       "8  chr17   7521600  7521800  chr17  7521788"
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#merge_regions(promp53, ['enhchr', 'enhstart', 'enhend'], mvp53, ['chrom', 'pos', 'pos'])\n",
    "#merge_regions(mvp53, ['chrom', 'pos', 'pos'], promp53, ['enhchr', 'enhstart', 'enhend'])\n",
    "#[df1, seqname1, start1, end1] = [promp53, 'enhchr', 'enhstart', 'enhend']\n",
    "#[df2, seqname2, start2, end2] = [mvp53, 'chrom', 'pos', 'pos']\n",
    "#c1 = df1.sort(end1).reset_index(drop=True)\n",
    "#c2 = df2.sort(start2).drop_duplicates(start2).reset_index(drop=True)\n",
    "#imerge = pd.cut(c1[end1], c2[start2], labels=False)\n",
    "keep = imerge >= 0\n",
    "cmerge = pd.concat([c1[['enhchr', 'enhstart', 'enhend']][keep], \n",
    "                    c2[['chrom', 'pos']].iloc[imerge[keep]].reset_index(drop=True)], axis=1).reset_index(drop=True)\n",
    "cmerge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10579583</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7521110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6915730</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7521767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12220598</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7521767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6915731</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7521770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10048648</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7521770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10136707</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7521770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6737809</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7521788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11741822</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7521788</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          chrom      pos\n",
       "10579583  chr17  7521110\n",
       "6915730   chr17  7521767\n",
       "12220598  chr17  7521767\n",
       "6915731   chr17  7521770\n",
       "10048648  chr17  7521770\n",
       "10136707  chr17  7521770\n",
       "6737809   chr17  7521788\n",
       "11741822  chr17  7521788"
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# find overlaps between ranges of two dataframes\n",
    "# return two arrays of indices into dataf\n",
    "def find_overlaps(df1, start1, end1, df2, start2, end2):\n",
    "    df1 = df1[[start1, end1]].sort(end1).reset_index(drop=True)\n",
    "    df2 = df2[[start2, end2]].sort(start2).reset_index(drop=True)\n",
    "    i1 = i2 = 0\n",
    "    result1 = []\n",
    "    result2 = []\n",
    "    while i1 < len(df1[start1]):\n",
    "        while "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  enhchr  enhstart   enhend\n",
      "0  chr17   7520200  7520400\n",
      "1  chr17   7520400  7520600\n",
      "2  chr17   7520600  7520800\n",
      "3  chr17   7520800  7521000\n",
      "4  chr17   7521000  7521200\n",
      "5  chr17   7521200  7521400\n",
      "6  chr17   7521400  7521600\n",
      "7  chr17   7521600  7521800\n",
      "8  chr17   7521600  7521800 \n",
      "    chrom      pos\n",
      "0   chr17  7519513\n",
      "1   chr17  7519769\n",
      "2   chr17  7519825\n",
      "3   chr17  7521110\n",
      "4   chr17  7521767\n",
      "5   chr17  7521770\n",
      "6   chr17  7521788\n",
      "7   chr17  7522135\n",
      "8   chr17  7522691\n",
      "9   chr17  7522708\n",
      "10  chr17  7523029\n",
      "11  chr17  7523438\n",
      "12  chr17  7523488 \n",
      "0    2\n",
      "1    2\n",
      "2    2\n",
      "3    2\n",
      "4    3\n",
      "5    3\n",
      "6    3\n",
      "7    6\n",
      "8    6\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print c1[['enhchr', 'enhstart', 'enhend']], \"\\n\", c2[['chrom', 'pos']], \"\\n\", imerge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3 2\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>enhchr</th>\n",
       "      <th>enhstart</th>\n",
       "      <th>enhend</th>\n",
       "      <th>CellLine</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>donor</th>\n",
       "      <th>Tumor_Seq_Allele_1</th>\n",
       "      <th>Tumor_Seq_Allele_2</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>tumorDNARatio</th>\n",
       "      <th>tumorRNARatio</th>\n",
       "      <th>FractionOfMeanExpression</th>\n",
       "      <th>zOfmeanExpression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7521000</td>\n",
       "      <td>7521200</td>\n",
       "      <td>Huvec</td>\n",
       "      <td>NM_000546</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7521110</td>\n",
       "      <td>0980e7fd-051d-45e9-9ca6-2baf073da4e8</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>0.232558</td>\n",
       "      <td>0.322581</td>\n",
       "      <td>0.195664</td>\n",
       "      <td>-1.617117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7521600</td>\n",
       "      <td>7521800</td>\n",
       "      <td>HSMM</td>\n",
       "      <td>NM_000546</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7521788</td>\n",
       "      <td>c612570a-881d-421f-98b9-0e4563133744</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>0.348837</td>\n",
       "      <td>2.527559</td>\n",
       "      <td>1.240533</td>\n",
       "      <td>0.406890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7521600</td>\n",
       "      <td>7521800</td>\n",
       "      <td>Huvec</td>\n",
       "      <td>NM_000546</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7521788</td>\n",
       "      <td>c612570a-881d-421f-98b9-0e4563133744</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>0.348837</td>\n",
       "      <td>2.527559</td>\n",
       "      <td>1.240533</td>\n",
       "      <td>0.406890</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  enhchr  enhstart   enhend CellLine  gene_name  chrom      pos  \\\n",
       "4  chr17   7521000  7521200    Huvec  NM_000546  chr17  7521110   \n",
       "7  chr17   7521600  7521800     HSMM  NM_000546  chr17  7521788   \n",
       "8  chr17   7521600  7521800    Huvec  NM_000546  chr17  7521788   \n",
       "\n",
       "                                  donor Tumor_Seq_Allele_1 Tumor_Seq_Allele_2  \\\n",
       "4  0980e7fd-051d-45e9-9ca6-2baf073da4e8                  G                  T   \n",
       "7  c612570a-881d-421f-98b9-0e4563133744                  T                  C   \n",
       "8  c612570a-881d-421f-98b9-0e4563133744                  T                  C   \n",
       "\n",
       "  Variant_Classification Variant_Type  tumorDNARatio  tumorRNARatio  \\\n",
       "4      Nonsense_Mutation          SNP       0.232558       0.322581   \n",
       "7      Missense_Mutation          SNP       0.348837       2.527559   \n",
       "8      Missense_Mutation          SNP       0.348837       2.527559   \n",
       "\n",
       "   FractionOfMeanExpression  zOfmeanExpression  \n",
       "4                  0.195664          -1.617117  \n",
       "7                  1.240533           0.406890  \n",
       "8                  1.240533           0.406890  "
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print len(mut_prom_vars.donor), len(set(mut_prom_vars.donor))\n",
    "mut_prom_vars[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>6737809</th>\n",
       "      <th>6915730</th>\n",
       "      <th>6915731</th>\n",
       "      <th>10048648</th>\n",
       "      <th>10136707</th>\n",
       "      <th>10579583</th>\n",
       "      <th>11741822</th>\n",
       "      <th>12220598</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RNAFile</th>\n",
       "      <td>stad\\6f62c5a4-97f9-4264-b633-edd368c44fb5-TP53...</td>\n",
       "      <td>brca\\12f66fb5-7444-410b-9c93-f275a09ce4dc-TP53...</td>\n",
       "      <td>brca\\12f66fb5-7444-410b-9c93-f275a09ce4dc-TP53...</td>\n",
       "      <td>brca\\3d40ae3e-c664-4370-95d8-780e760d6d08-TP53...</td>\n",
       "      <td>brca\\6e9d893d-fdf1-45b4-b1e0-7142863aafe9-TP53...</td>\n",
       "      <td>coad\\192a5602-3bc2-4bcc-8a87-ef039626f092-TP53...</td>\n",
       "      <td>stad\\4fc386ae-5646-4041-ac9a-0747d3162d26-TP53...</td>\n",
       "      <td>lusc\\548da644-a577-4c86-b04e-2ca21fe4c774-TP53...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNAFile</th>\n",
       "      <td>stad\\ada589d6-c2db-42c0-a927-7ddf0cb3de85-TP53...</td>\n",
       "      <td>brca\\8f0ce179-a754-464d-84b6-31f8a1a8f77f-TP53...</td>\n",
       "      <td>brca\\8f0ce179-a754-464d-84b6-31f8a1a8f77f-TP53...</td>\n",
       "      <td>brca\\f945990a-00d4-4cef-848b-8f145dd27601-TP53...</td>\n",
       "      <td>brca\\649630d6-3318-4e29-b003-71980bbde398-TP53...</td>\n",
       "      <td>coad\\314d37dc-625e-4f01-87a3-4a6a884bd32f-TP53...</td>\n",
       "      <td>stad\\8d796323-4cea-4633-86e8-bc4f3fb0b3c1-TP53...</td>\n",
       "      <td>lusc\\1ccbfc36-db90-477a-9333-c733b12ff52c-TP53...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <td>TP53</td>\n",
       "      <td>TP53</td>\n",
       "      <td>TP53</td>\n",
       "      <td>TP53</td>\n",
       "      <td>TP53</td>\n",
       "      <td>TP53</td>\n",
       "      <td>TP53</td>\n",
       "      <td>TP53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <td>7157</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7157</td>\n",
       "      <td>7157</td>\n",
       "      <td>7157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Center</th>\n",
       "      <td>broad.mit.edu</td>\n",
       "      <td>genome.wustl.edu</td>\n",
       "      <td>genome.wustl.edu</td>\n",
       "      <td>genome.wustl.edu</td>\n",
       "      <td>genome.wustl.edu</td>\n",
       "      <td>hgsc.bcm.edu</td>\n",
       "      <td>broad.mit.edu</td>\n",
       "      <td>broad.mit.edu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NCBI_Build</th>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Chromosome</th>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Start_Position</th>\n",
       "      <td>7578271</td>\n",
       "      <td>7578403</td>\n",
       "      <td>7578403</td>\n",
       "      <td>7578188</td>\n",
       "      <td>7578212</td>\n",
       "      <td>7579406</td>\n",
       "      <td>7578424</td>\n",
       "      <td>7579470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>End_Position</th>\n",
       "      <td>7578271</td>\n",
       "      <td>7578403</td>\n",
       "      <td>7578403</td>\n",
       "      <td>7578188</td>\n",
       "      <td>7578212</td>\n",
       "      <td>7579406</td>\n",
       "      <td>7578424</td>\n",
       "      <td>7579470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Strand</th>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Variant_Classification</th>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Variant_Type</th>\n",
       "      <td>SNP</td>\n",
       "      <td>SNP</td>\n",
       "      <td>SNP</td>\n",
       "      <td>SNP</td>\n",
       "      <td>SNP</td>\n",
       "      <td>SNP</td>\n",
       "      <td>SNP</td>\n",
       "      <td>DEL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Reference_Allele</th>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tumor_Seq_Allele_1</th>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tumor_Seq_Allele_2</th>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dbSNP_RS</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>novel</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dbSNP_Val_Status</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <td>TCGA-BR-6456-01A-11D-1800-08</td>\n",
       "      <td>TCGA-GI-A2C9-01A-11D-A21Q-09</td>\n",
       "      <td>TCGA-GI-A2C9-01A-11D-A21Q-09</td>\n",
       "      <td>TCGA-A7-A13D-01A-13D-A12Q-09</td>\n",
       "      <td>TCGA-AO-A124-01A-11D-A10M-09</td>\n",
       "      <td>TCGA-CA-6718-01A-11D-1835-10</td>\n",
       "      <td>TCGA-CG-5723-01A-11D-1600-08</td>\n",
       "      <td>TCGA-60-2713-01A-01D-1522-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Matched_Norm_Sample_Barcode</th>\n",
       "      <td>TCGA-BR-6456-10A-01D-1800-08</td>\n",
       "      <td>TCGA-GI-A2C9-11A-22D-A21Q-09</td>\n",
       "      <td>TCGA-GI-A2C9-11A-22D-A21Q-09</td>\n",
       "      <td>TCGA-A7-A13D-10A-02D-A12Q-09</td>\n",
       "      <td>TCGA-AO-A124-10A-01D-A10M-09</td>\n",
       "      <td>TCGA-CA-6718-10A-01D-1835-10</td>\n",
       "      <td>TCGA-CG-5723-11A-02D-1600-08</td>\n",
       "      <td>TCGA-60-2713-11A-01D-1522-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Match_Norm_Seq_Allele1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Match_Norm_Seq_Allele2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tumor_Validation_Allele1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tumor_Validation_Allele2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Match_Norm_Validation_Allele1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Match_Norm_Validation_Allele2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Verification_Status</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Validation_Status</th>\n",
       "      <td>Untested</td>\n",
       "      <td>Untested</td>\n",
       "      <td>Untested</td>\n",
       "      <td>Untested</td>\n",
       "      <td>Untested</td>\n",
       "      <td>Untested</td>\n",
       "      <td>Untested</td>\n",
       "      <td>Untested</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mutation_Status</th>\n",
       "      <td>Somatic</td>\n",
       "      <td>Somatic</td>\n",
       "      <td>Somatic</td>\n",
       "      <td>Somatic</td>\n",
       "      <td>Somatic</td>\n",
       "      <td>Somatic</td>\n",
       "      <td>Somatic</td>\n",
       "      <td>Somatic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sequencing_Phase</th>\n",
       "      <td>Phase_I</td>\n",
       "      <td>Phase_IV</td>\n",
       "      <td>Phase_IV</td>\n",
       "      <td>Phase_IV</td>\n",
       "      <td>Phase_IV</td>\n",
       "      <td>Phase_I</td>\n",
       "      <td>Phase_I</td>\n",
       "      <td>Phase_I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sequence_Source</th>\n",
       "      <td>WXS</td>\n",
       "      <td>WXS</td>\n",
       "      <td>WXS</td>\n",
       "      <td>WXS</td>\n",
       "      <td>WXS</td>\n",
       "      <td>WXS</td>\n",
       "      <td>WXS</td>\n",
       "      <td>WXS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>n_DNA_Matching_Both</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>n_RNA_Matching_Reference</th>\n",
       "      <td>117</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>18</td>\n",
       "      <td>6</td>\n",
       "      <td>28</td>\n",
       "      <td>83</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>n_RNA_Matching_Tumor</th>\n",
       "      <td>321</td>\n",
       "      <td>181</td>\n",
       "      <td>181</td>\n",
       "      <td>4</td>\n",
       "      <td>22</td>\n",
       "      <td>10</td>\n",
       "      <td>36</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>n_RNA_Matching_Neither</th>\n",
       "      <td>10</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>12</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>n_RNA_Matching_Both</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tumorDNAFraction</th>\n",
       "      <td>0.2586207</td>\n",
       "      <td>0.6181818</td>\n",
       "      <td>0.6181818</td>\n",
       "      <td>0.6973684</td>\n",
       "      <td>0.7179487</td>\n",
       "      <td>0.1886792</td>\n",
       "      <td>0.1460674</td>\n",
       "      <td>0.03448276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tumorRNAFraction</th>\n",
       "      <td>0.7165179</td>\n",
       "      <td>0.8743961</td>\n",
       "      <td>0.8743961</td>\n",
       "      <td>0.173913</td>\n",
       "      <td>0.7586207</td>\n",
       "      <td>0.2439024</td>\n",
       "      <td>0.2748092</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tumorDNAMultiple</th>\n",
       "      <td>-2.866667</td>\n",
       "      <td>1.619048</td>\n",
       "      <td>1.619048</td>\n",
       "      <td>2.304348</td>\n",
       "      <td>2.545455</td>\n",
       "      <td>-4.3</td>\n",
       "      <td>-5.846154</td>\n",
       "      <td>-28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tumorRNAMultiple</th>\n",
       "      <td>2.527559</td>\n",
       "      <td>6.961538</td>\n",
       "      <td>6.961538</td>\n",
       "      <td>-4.75</td>\n",
       "      <td>3.142857</td>\n",
       "      <td>-3.1</td>\n",
       "      <td>-2.638889</td>\n",
       "      <td>-30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tumorDNARatio</th>\n",
       "      <td>0.3488372</td>\n",
       "      <td>1.619048</td>\n",
       "      <td>1.619048</td>\n",
       "      <td>2.304348</td>\n",
       "      <td>2.545455</td>\n",
       "      <td>0.2325581</td>\n",
       "      <td>0.1710526</td>\n",
       "      <td>0.03571429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tumorRNARatio</th>\n",
       "      <td>2.527559</td>\n",
       "      <td>6.961538</td>\n",
       "      <td>6.961538</td>\n",
       "      <td>0.2105263</td>\n",
       "      <td>3.142857</td>\n",
       "      <td>0.3225806</td>\n",
       "      <td>0.3789474</td>\n",
       "      <td>0.0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FractionOfMeanExpression</th>\n",
       "      <td>1.240533</td>\n",
       "      <td>1.353312</td>\n",
       "      <td>1.353312</td>\n",
       "      <td>0.2492425</td>\n",
       "      <td>0.1690905</td>\n",
       "      <td>0.1956636</td>\n",
       "      <td>1.139024</td>\n",
       "      <td>0.3217284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zOfmeanExpression</th>\n",
       "      <td>0.4068902</td>\n",
       "      <td>0.6485612</td>\n",
       "      <td>0.6485612</td>\n",
       "      <td>-1.378136</td>\n",
       "      <td>-1.525268</td>\n",
       "      <td>-1.617117</td>\n",
       "      <td>0.235176</td>\n",
       "      <td>-0.8527172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IsSingle</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CancerType</th>\n",
       "      <td>stad</td>\n",
       "      <td>brca</td>\n",
       "      <td>brca</td>\n",
       "      <td>brca</td>\n",
       "      <td>brca</td>\n",
       "      <td>coad</td>\n",
       "      <td>stad</td>\n",
       "      <td>lusc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gender</th>\n",
       "      <td>female</td>\n",
       "      <td>female</td>\n",
       "      <td>female</td>\n",
       "      <td>female</td>\n",
       "      <td>female</td>\n",
       "      <td>male</td>\n",
       "      <td>male</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <td>TP53</td>\n",
       "      <td>TP53</td>\n",
       "      <td>TP53</td>\n",
       "      <td>TP53</td>\n",
       "      <td>TP53</td>\n",
       "      <td>TP53</td>\n",
       "      <td>TP53</td>\n",
       "      <td>TP53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>start</th>\n",
       "      <td>7578271</td>\n",
       "      <td>7578403</td>\n",
       "      <td>7578403</td>\n",
       "      <td>7578188</td>\n",
       "      <td>7578212</td>\n",
       "      <td>7579406</td>\n",
       "      <td>7578424</td>\n",
       "      <td>7579470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>end</th>\n",
       "      <td>7578271</td>\n",
       "      <td>7578403</td>\n",
       "      <td>7578403</td>\n",
       "      <td>7578188</td>\n",
       "      <td>7578212</td>\n",
       "      <td>7579406</td>\n",
       "      <td>7578424</td>\n",
       "      <td>7579470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr</th>\n",
       "      <td>chr17</td>\n",
       "      <td>chr17</td>\n",
       "      <td>chr17</td>\n",
       "      <td>chr17</td>\n",
       "      <td>chr17</td>\n",
       "      <td>chr17</td>\n",
       "      <td>chr17</td>\n",
       "      <td>chr17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_allele</th>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrom</th>\n",
       "      <td>chr17</td>\n",
       "      <td>chr17</td>\n",
       "      <td>chr17</td>\n",
       "      <td>chr17</td>\n",
       "      <td>chr17</td>\n",
       "      <td>chr17</td>\n",
       "      <td>chr17</td>\n",
       "      <td>chr17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pos</th>\n",
       "      <td>7521788</td>\n",
       "      <td>7521767</td>\n",
       "      <td>7521770</td>\n",
       "      <td>7521770</td>\n",
       "      <td>7521770</td>\n",
       "      <td>7521110</td>\n",
       "      <td>7521788</td>\n",
       "      <td>7521767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <td>bf63e18c-9dfe-11e4-bbc8-af52830fd8cf</td>\n",
       "      <td>677c6d7a-dd66-11e4-9a3a-d51a9602cc95</td>\n",
       "      <td>677c6e38-dd66-11e4-9a3a-d51a9602cc95</td>\n",
       "      <td>79aeb56e-dda4-11e4-88a0-6b3b9fbecd02</td>\n",
       "      <td>8c187296-dda2-11e4-9871-805ff2276008</td>\n",
       "      <td>d5010cb8-c603-11e4-95fd-0cf8e19d4dc7</td>\n",
       "      <td>7d61ad22-c96f-11e4-936d-4929716f0807</td>\n",
       "      <td>7cd88364-ceef-11e4-86db-c482ec69b06c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ref</th>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>alt</th>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>qual</th>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>filter</th>\n",
       "      <td>SR;MN;RP;VUM;MNP</td>\n",
       "      <td>SR;RP;VUM</td>\n",
       "      <td>SR;MN;VUM;MNP</td>\n",
       "      <td>SR;MN;VUM;MNP</td>\n",
       "      <td>SR;VUM</td>\n",
       "      <td>PASS</td>\n",
       "      <td>SR;RP;VUM</td>\n",
       "      <td>SR;MN;RP;VUM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>normal</th>\n",
       "      <td>0|0:0:0:19:0:0:0:12:2:6.1e-02</td>\n",
       "      <td>0|0:0:0:0:1:0:0:0:8:0.0e+00</td>\n",
       "      <td>0|0:0:0:1:0:0:0:8:0:0.0e+00</td>\n",
       "      <td>0|0:0:0:3:0:0:0:4:0:0.0e+00</td>\n",
       "      <td>0|0:0:0:6:0:0:0:14:0:0.0e+00</td>\n",
       "      <td>0|0:23:0:0:0:10:0:0:0:0.0e+00</td>\n",
       "      <td>0|0:0:0:15:0:0:0:2:0:0.0e+00</td>\n",
       "      <td>0|0:0:0:0:5:0:0:0:17:0.0e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tumor</th>\n",
       "      <td>0|1:0:0:27:2:0:0:11:2:9.5e-02</td>\n",
       "      <td>0|1:0:0:4:5:0:0:0:10:2.1e-01</td>\n",
       "      <td>0|1:0:0:4:4:0:0:9:0:2.4e-01</td>\n",
       "      <td>0|1:0:0:12:5:0:0:8:0:2.0e-01</td>\n",
       "      <td>0|1:0:0:3:3:0:0:9:1:2.5e-01</td>\n",
       "      <td>0|1:28:13:0:0:21:6:0:0:2.8e-01</td>\n",
       "      <td>0|1:0:0:20:3:0:0:14:2:1.3e-01</td>\n",
       "      <td>0|1:0:0:3:1:0:0:0:6:3.0e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>72 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                        6737809   \\\n",
       "RNAFile                        stad\\6f62c5a4-97f9-4264-b633-edd368c44fb5-TP53...   \n",
       "DNAFile                        stad\\ada589d6-c2db-42c0-a927-7ddf0cb3de85-TP53...   \n",
       "Hugo_Symbol                                                                 TP53   \n",
       "Entrez_Gene_Id                                                              7157   \n",
       "Center                                                             broad.mit.edu   \n",
       "NCBI_Build                                                                    37   \n",
       "Chromosome                                                                    17   \n",
       "Start_Position                                                           7578271   \n",
       "End_Position                                                             7578271   \n",
       "Strand                                                                         +   \n",
       "Variant_Classification                                         Missense_Mutation   \n",
       "Variant_Type                                                                 SNP   \n",
       "Reference_Allele                                                               T   \n",
       "Tumor_Seq_Allele_1                                                             T   \n",
       "Tumor_Seq_Allele_2                                                             C   \n",
       "dbSNP_RS                                                                     NaN   \n",
       "dbSNP_Val_Status                                                             NaN   \n",
       "Tumor_Sample_Barcode                                TCGA-BR-6456-01A-11D-1800-08   \n",
       "Matched_Norm_Sample_Barcode                         TCGA-BR-6456-10A-01D-1800-08   \n",
       "Match_Norm_Seq_Allele1                                                       NaN   \n",
       "Match_Norm_Seq_Allele2                                                       NaN   \n",
       "Tumor_Validation_Allele1                                                     NaN   \n",
       "Tumor_Validation_Allele2                                                     NaN   \n",
       "Match_Norm_Validation_Allele1                                                NaN   \n",
       "Match_Norm_Validation_Allele2                                                NaN   \n",
       "Verification_Status                                                          NaN   \n",
       "Validation_Status                                                       Untested   \n",
       "Mutation_Status                                                          Somatic   \n",
       "Sequencing_Phase                                                         Phase_I   \n",
       "Sequence_Source                                                              WXS   \n",
       "...                                                                          ...   \n",
       "n_DNA_Matching_Both                                                            0   \n",
       "n_RNA_Matching_Reference                                                     117   \n",
       "n_RNA_Matching_Tumor                                                         321   \n",
       "n_RNA_Matching_Neither                                                        10   \n",
       "n_RNA_Matching_Both                                                            0   \n",
       "tumorDNAFraction                                                       0.2586207   \n",
       "tumorRNAFraction                                                       0.7165179   \n",
       "tumorDNAMultiple                                                       -2.866667   \n",
       "tumorRNAMultiple                                                        2.527559   \n",
       "tumorDNARatio                                                          0.3488372   \n",
       "tumorRNARatio                                                           2.527559   \n",
       "FractionOfMeanExpression                                                1.240533   \n",
       "zOfmeanExpression                                                      0.4068902   \n",
       "IsSingle                                                                    True   \n",
       "CancerType                                                                  stad   \n",
       "gender                                                                    female   \n",
       "gene                                                                        TP53   \n",
       "start                                                                    7578271   \n",
       "end                                                                      7578271   \n",
       "chr                                                                        chr17   \n",
       "var_allele                                                                     T   \n",
       "chrom                                                                      chr17   \n",
       "pos                                                                      7521788   \n",
       "id                                          bf63e18c-9dfe-11e4-bbc8-af52830fd8cf   \n",
       "ref                                                                            G   \n",
       "alt                                                                            T   \n",
       "qual                                                                           .   \n",
       "filter                                                          SR;MN;RP;VUM;MNP   \n",
       "normal                                             0|0:0:0:19:0:0:0:12:2:6.1e-02   \n",
       "tumor                                              0|1:0:0:27:2:0:0:11:2:9.5e-02   \n",
       "\n",
       "                                                                        6915730   \\\n",
       "RNAFile                        brca\\12f66fb5-7444-410b-9c93-f275a09ce4dc-TP53...   \n",
       "DNAFile                        brca\\8f0ce179-a754-464d-84b6-31f8a1a8f77f-TP53...   \n",
       "Hugo_Symbol                                                                 TP53   \n",
       "Entrez_Gene_Id                                                                 0   \n",
       "Center                                                          genome.wustl.edu   \n",
       "NCBI_Build                                                                    37   \n",
       "Chromosome                                                                    17   \n",
       "Start_Position                                                           7578403   \n",
       "End_Position                                                             7578403   \n",
       "Strand                                                                         +   \n",
       "Variant_Classification                                         Missense_Mutation   \n",
       "Variant_Type                                                                 SNP   \n",
       "Reference_Allele                                                               C   \n",
       "Tumor_Seq_Allele_1                                                             C   \n",
       "Tumor_Seq_Allele_2                                                             T   \n",
       "dbSNP_RS                                                                     NaN   \n",
       "dbSNP_Val_Status                                                             NaN   \n",
       "Tumor_Sample_Barcode                                TCGA-GI-A2C9-01A-11D-A21Q-09   \n",
       "Matched_Norm_Sample_Barcode                         TCGA-GI-A2C9-11A-22D-A21Q-09   \n",
       "Match_Norm_Seq_Allele1                                                         C   \n",
       "Match_Norm_Seq_Allele2                                                         C   \n",
       "Tumor_Validation_Allele1                                                     NaN   \n",
       "Tumor_Validation_Allele2                                                     NaN   \n",
       "Match_Norm_Validation_Allele1                                                NaN   \n",
       "Match_Norm_Validation_Allele2                                                NaN   \n",
       "Verification_Status                                                      Unknown   \n",
       "Validation_Status                                                       Untested   \n",
       "Mutation_Status                                                          Somatic   \n",
       "Sequencing_Phase                                                        Phase_IV   \n",
       "Sequence_Source                                                              WXS   \n",
       "...                                                                          ...   \n",
       "n_DNA_Matching_Both                                                            0   \n",
       "n_RNA_Matching_Reference                                                      13   \n",
       "n_RNA_Matching_Tumor                                                         181   \n",
       "n_RNA_Matching_Neither                                                        13   \n",
       "n_RNA_Matching_Both                                                            0   \n",
       "tumorDNAFraction                                                       0.6181818   \n",
       "tumorRNAFraction                                                       0.8743961   \n",
       "tumorDNAMultiple                                                        1.619048   \n",
       "tumorRNAMultiple                                                        6.961538   \n",
       "tumorDNARatio                                                           1.619048   \n",
       "tumorRNARatio                                                           6.961538   \n",
       "FractionOfMeanExpression                                                1.353312   \n",
       "zOfmeanExpression                                                      0.6485612   \n",
       "IsSingle                                                                    True   \n",
       "CancerType                                                                  brca   \n",
       "gender                                                                    female   \n",
       "gene                                                                        TP53   \n",
       "start                                                                    7578403   \n",
       "end                                                                      7578403   \n",
       "chr                                                                        chr17   \n",
       "var_allele                                                                     C   \n",
       "chrom                                                                      chr17   \n",
       "pos                                                                      7521767   \n",
       "id                                          677c6d7a-dd66-11e4-9a3a-d51a9602cc95   \n",
       "ref                                                                            T   \n",
       "alt                                                                            G   \n",
       "qual                                                                           .   \n",
       "filter                                                                 SR;RP;VUM   \n",
       "normal                                               0|0:0:0:0:1:0:0:0:8:0.0e+00   \n",
       "tumor                                               0|1:0:0:4:5:0:0:0:10:2.1e-01   \n",
       "\n",
       "                                                                        6915731   \\\n",
       "RNAFile                        brca\\12f66fb5-7444-410b-9c93-f275a09ce4dc-TP53...   \n",
       "DNAFile                        brca\\8f0ce179-a754-464d-84b6-31f8a1a8f77f-TP53...   \n",
       "Hugo_Symbol                                                                 TP53   \n",
       "Entrez_Gene_Id                                                                 0   \n",
       "Center                                                          genome.wustl.edu   \n",
       "NCBI_Build                                                                    37   \n",
       "Chromosome                                                                    17   \n",
       "Start_Position                                                           7578403   \n",
       "End_Position                                                             7578403   \n",
       "Strand                                                                         +   \n",
       "Variant_Classification                                         Missense_Mutation   \n",
       "Variant_Type                                                                 SNP   \n",
       "Reference_Allele                                                               C   \n",
       "Tumor_Seq_Allele_1                                                             C   \n",
       "Tumor_Seq_Allele_2                                                             T   \n",
       "dbSNP_RS                                                                     NaN   \n",
       "dbSNP_Val_Status                                                             NaN   \n",
       "Tumor_Sample_Barcode                                TCGA-GI-A2C9-01A-11D-A21Q-09   \n",
       "Matched_Norm_Sample_Barcode                         TCGA-GI-A2C9-11A-22D-A21Q-09   \n",
       "Match_Norm_Seq_Allele1                                                         C   \n",
       "Match_Norm_Seq_Allele2                                                         C   \n",
       "Tumor_Validation_Allele1                                                     NaN   \n",
       "Tumor_Validation_Allele2                                                     NaN   \n",
       "Match_Norm_Validation_Allele1                                                NaN   \n",
       "Match_Norm_Validation_Allele2                                                NaN   \n",
       "Verification_Status                                                      Unknown   \n",
       "Validation_Status                                                       Untested   \n",
       "Mutation_Status                                                          Somatic   \n",
       "Sequencing_Phase                                                        Phase_IV   \n",
       "Sequence_Source                                                              WXS   \n",
       "...                                                                          ...   \n",
       "n_DNA_Matching_Both                                                            0   \n",
       "n_RNA_Matching_Reference                                                      13   \n",
       "n_RNA_Matching_Tumor                                                         181   \n",
       "n_RNA_Matching_Neither                                                        13   \n",
       "n_RNA_Matching_Both                                                            0   \n",
       "tumorDNAFraction                                                       0.6181818   \n",
       "tumorRNAFraction                                                       0.8743961   \n",
       "tumorDNAMultiple                                                        1.619048   \n",
       "tumorRNAMultiple                                                        6.961538   \n",
       "tumorDNARatio                                                           1.619048   \n",
       "tumorRNARatio                                                           6.961538   \n",
       "FractionOfMeanExpression                                                1.353312   \n",
       "zOfmeanExpression                                                      0.6485612   \n",
       "IsSingle                                                                    True   \n",
       "CancerType                                                                  brca   \n",
       "gender                                                                    female   \n",
       "gene                                                                        TP53   \n",
       "start                                                                    7578403   \n",
       "end                                                                      7578403   \n",
       "chr                                                                        chr17   \n",
       "var_allele                                                                     C   \n",
       "chrom                                                                      chr17   \n",
       "pos                                                                      7521770   \n",
       "id                                          677c6e38-dd66-11e4-9a3a-d51a9602cc95   \n",
       "ref                                                                            G   \n",
       "alt                                                                            T   \n",
       "qual                                                                           .   \n",
       "filter                                                             SR;MN;VUM;MNP   \n",
       "normal                                               0|0:0:0:1:0:0:0:8:0:0.0e+00   \n",
       "tumor                                                0|1:0:0:4:4:0:0:9:0:2.4e-01   \n",
       "\n",
       "                                                                        10048648  \\\n",
       "RNAFile                        brca\\3d40ae3e-c664-4370-95d8-780e760d6d08-TP53...   \n",
       "DNAFile                        brca\\f945990a-00d4-4cef-848b-8f145dd27601-TP53...   \n",
       "Hugo_Symbol                                                                 TP53   \n",
       "Entrez_Gene_Id                                                                 0   \n",
       "Center                                                          genome.wustl.edu   \n",
       "NCBI_Build                                                                    37   \n",
       "Chromosome                                                                    17   \n",
       "Start_Position                                                           7578188   \n",
       "End_Position                                                             7578188   \n",
       "Strand                                                                         +   \n",
       "Variant_Classification                                         Nonsense_Mutation   \n",
       "Variant_Type                                                                 SNP   \n",
       "Reference_Allele                                                               C   \n",
       "Tumor_Seq_Allele_1                                                             C   \n",
       "Tumor_Seq_Allele_2                                                             A   \n",
       "dbSNP_RS                                                                     NaN   \n",
       "dbSNP_Val_Status                                                             NaN   \n",
       "Tumor_Sample_Barcode                                TCGA-A7-A13D-01A-13D-A12Q-09   \n",
       "Matched_Norm_Sample_Barcode                         TCGA-A7-A13D-10A-02D-A12Q-09   \n",
       "Match_Norm_Seq_Allele1                                                         C   \n",
       "Match_Norm_Seq_Allele2                                                         C   \n",
       "Tumor_Validation_Allele1                                                     NaN   \n",
       "Tumor_Validation_Allele2                                                     NaN   \n",
       "Match_Norm_Validation_Allele1                                                NaN   \n",
       "Match_Norm_Validation_Allele2                                                NaN   \n",
       "Verification_Status                                                      Unknown   \n",
       "Validation_Status                                                       Untested   \n",
       "Mutation_Status                                                          Somatic   \n",
       "Sequencing_Phase                                                        Phase_IV   \n",
       "Sequence_Source                                                              WXS   \n",
       "...                                                                          ...   \n",
       "n_DNA_Matching_Both                                                            0   \n",
       "n_RNA_Matching_Reference                                                      18   \n",
       "n_RNA_Matching_Tumor                                                           4   \n",
       "n_RNA_Matching_Neither                                                         1   \n",
       "n_RNA_Matching_Both                                                            0   \n",
       "tumorDNAFraction                                                       0.6973684   \n",
       "tumorRNAFraction                                                        0.173913   \n",
       "tumorDNAMultiple                                                        2.304348   \n",
       "tumorRNAMultiple                                                           -4.75   \n",
       "tumorDNARatio                                                           2.304348   \n",
       "tumorRNARatio                                                          0.2105263   \n",
       "FractionOfMeanExpression                                               0.2492425   \n",
       "zOfmeanExpression                                                      -1.378136   \n",
       "IsSingle                                                                    True   \n",
       "CancerType                                                                  brca   \n",
       "gender                                                                    female   \n",
       "gene                                                                        TP53   \n",
       "start                                                                    7578188   \n",
       "end                                                                      7578188   \n",
       "chr                                                                        chr17   \n",
       "var_allele                                                                     C   \n",
       "chrom                                                                      chr17   \n",
       "pos                                                                      7521770   \n",
       "id                                          79aeb56e-dda4-11e4-88a0-6b3b9fbecd02   \n",
       "ref                                                                            G   \n",
       "alt                                                                            T   \n",
       "qual                                                                           .   \n",
       "filter                                                             SR;MN;VUM;MNP   \n",
       "normal                                               0|0:0:0:3:0:0:0:4:0:0.0e+00   \n",
       "tumor                                               0|1:0:0:12:5:0:0:8:0:2.0e-01   \n",
       "\n",
       "                                                                        10136707  \\\n",
       "RNAFile                        brca\\6e9d893d-fdf1-45b4-b1e0-7142863aafe9-TP53...   \n",
       "DNAFile                        brca\\649630d6-3318-4e29-b003-71980bbde398-TP53...   \n",
       "Hugo_Symbol                                                                 TP53   \n",
       "Entrez_Gene_Id                                                                 0   \n",
       "Center                                                          genome.wustl.edu   \n",
       "NCBI_Build                                                                    37   \n",
       "Chromosome                                                                    17   \n",
       "Start_Position                                                           7578212   \n",
       "End_Position                                                             7578212   \n",
       "Strand                                                                         +   \n",
       "Variant_Classification                                         Nonsense_Mutation   \n",
       "Variant_Type                                                                 SNP   \n",
       "Reference_Allele                                                               G   \n",
       "Tumor_Seq_Allele_1                                                             G   \n",
       "Tumor_Seq_Allele_2                                                             A   \n",
       "dbSNP_RS                                                                     NaN   \n",
       "dbSNP_Val_Status                                                             NaN   \n",
       "Tumor_Sample_Barcode                                TCGA-AO-A124-01A-11D-A10M-09   \n",
       "Matched_Norm_Sample_Barcode                         TCGA-AO-A124-10A-01D-A10M-09   \n",
       "Match_Norm_Seq_Allele1                                                         G   \n",
       "Match_Norm_Seq_Allele2                                                         G   \n",
       "Tumor_Validation_Allele1                                                     NaN   \n",
       "Tumor_Validation_Allele2                                                     NaN   \n",
       "Match_Norm_Validation_Allele1                                                NaN   \n",
       "Match_Norm_Validation_Allele2                                                NaN   \n",
       "Verification_Status                                                      Unknown   \n",
       "Validation_Status                                                       Untested   \n",
       "Mutation_Status                                                          Somatic   \n",
       "Sequencing_Phase                                                        Phase_IV   \n",
       "Sequence_Source                                                              WXS   \n",
       "...                                                                          ...   \n",
       "n_DNA_Matching_Both                                                            0   \n",
       "n_RNA_Matching_Reference                                                       6   \n",
       "n_RNA_Matching_Tumor                                                          22   \n",
       "n_RNA_Matching_Neither                                                         1   \n",
       "n_RNA_Matching_Both                                                            0   \n",
       "tumorDNAFraction                                                       0.7179487   \n",
       "tumorRNAFraction                                                       0.7586207   \n",
       "tumorDNAMultiple                                                        2.545455   \n",
       "tumorRNAMultiple                                                        3.142857   \n",
       "tumorDNARatio                                                           2.545455   \n",
       "tumorRNARatio                                                           3.142857   \n",
       "FractionOfMeanExpression                                               0.1690905   \n",
       "zOfmeanExpression                                                      -1.525268   \n",
       "IsSingle                                                                    True   \n",
       "CancerType                                                                  brca   \n",
       "gender                                                                    female   \n",
       "gene                                                                        TP53   \n",
       "start                                                                    7578212   \n",
       "end                                                                      7578212   \n",
       "chr                                                                        chr17   \n",
       "var_allele                                                                     G   \n",
       "chrom                                                                      chr17   \n",
       "pos                                                                      7521770   \n",
       "id                                          8c187296-dda2-11e4-9871-805ff2276008   \n",
       "ref                                                                            G   \n",
       "alt                                                                            T   \n",
       "qual                                                                           .   \n",
       "filter                                                                    SR;VUM   \n",
       "normal                                              0|0:0:0:6:0:0:0:14:0:0.0e+00   \n",
       "tumor                                                0|1:0:0:3:3:0:0:9:1:2.5e-01   \n",
       "\n",
       "                                                                        10579583  \\\n",
       "RNAFile                        coad\\192a5602-3bc2-4bcc-8a87-ef039626f092-TP53...   \n",
       "DNAFile                        coad\\314d37dc-625e-4f01-87a3-4a6a884bd32f-TP53...   \n",
       "Hugo_Symbol                                                                 TP53   \n",
       "Entrez_Gene_Id                                                              7157   \n",
       "Center                                                              hgsc.bcm.edu   \n",
       "NCBI_Build                                                                    37   \n",
       "Chromosome                                                                    17   \n",
       "Start_Position                                                           7579406   \n",
       "End_Position                                                             7579406   \n",
       "Strand                                                                         +   \n",
       "Variant_Classification                                         Nonsense_Mutation   \n",
       "Variant_Type                                                                 SNP   \n",
       "Reference_Allele                                                               G   \n",
       "Tumor_Seq_Allele_1                                                             G   \n",
       "Tumor_Seq_Allele_2                                                             T   \n",
       "dbSNP_RS                                                                   novel   \n",
       "dbSNP_Val_Status                                                             NaN   \n",
       "Tumor_Sample_Barcode                                TCGA-CA-6718-01A-11D-1835-10   \n",
       "Matched_Norm_Sample_Barcode                         TCGA-CA-6718-10A-01D-1835-10   \n",
       "Match_Norm_Seq_Allele1                                                         G   \n",
       "Match_Norm_Seq_Allele2                                                         G   \n",
       "Tumor_Validation_Allele1                                                       G   \n",
       "Tumor_Validation_Allele2                                                       T   \n",
       "Match_Norm_Validation_Allele1                                                  G   \n",
       "Match_Norm_Validation_Allele2                                                  G   \n",
       "Verification_Status                                                      Unknown   \n",
       "Validation_Status                                                       Untested   \n",
       "Mutation_Status                                                          Somatic   \n",
       "Sequencing_Phase                                                         Phase_I   \n",
       "Sequence_Source                                                              WXS   \n",
       "...                                                                          ...   \n",
       "n_DNA_Matching_Both                                                            0   \n",
       "n_RNA_Matching_Reference                                                      28   \n",
       "n_RNA_Matching_Tumor                                                          10   \n",
       "n_RNA_Matching_Neither                                                         3   \n",
       "n_RNA_Matching_Both                                                            0   \n",
       "tumorDNAFraction                                                       0.1886792   \n",
       "tumorRNAFraction                                                       0.2439024   \n",
       "tumorDNAMultiple                                                            -4.3   \n",
       "tumorRNAMultiple                                                            -3.1   \n",
       "tumorDNARatio                                                          0.2325581   \n",
       "tumorRNARatio                                                          0.3225806   \n",
       "FractionOfMeanExpression                                               0.1956636   \n",
       "zOfmeanExpression                                                      -1.617117   \n",
       "IsSingle                                                                    True   \n",
       "CancerType                                                                  coad   \n",
       "gender                                                                      male   \n",
       "gene                                                                        TP53   \n",
       "start                                                                    7579406   \n",
       "end                                                                      7579406   \n",
       "chr                                                                        chr17   \n",
       "var_allele                                                                     G   \n",
       "chrom                                                                      chr17   \n",
       "pos                                                                      7521110   \n",
       "id                                          d5010cb8-c603-11e4-95fd-0cf8e19d4dc7   \n",
       "ref                                                                            A   \n",
       "alt                                                                            C   \n",
       "qual                                                                           .   \n",
       "filter                                                                      PASS   \n",
       "normal                                             0|0:23:0:0:0:10:0:0:0:0.0e+00   \n",
       "tumor                                             0|1:28:13:0:0:21:6:0:0:2.8e-01   \n",
       "\n",
       "                                                                        11741822  \\\n",
       "RNAFile                        stad\\4fc386ae-5646-4041-ac9a-0747d3162d26-TP53...   \n",
       "DNAFile                        stad\\8d796323-4cea-4633-86e8-bc4f3fb0b3c1-TP53...   \n",
       "Hugo_Symbol                                                                 TP53   \n",
       "Entrez_Gene_Id                                                              7157   \n",
       "Center                                                             broad.mit.edu   \n",
       "NCBI_Build                                                                    37   \n",
       "Chromosome                                                                    17   \n",
       "Start_Position                                                           7578424   \n",
       "End_Position                                                             7578424   \n",
       "Strand                                                                         +   \n",
       "Variant_Classification                                         Missense_Mutation   \n",
       "Variant_Type                                                                 SNP   \n",
       "Reference_Allele                                                               A   \n",
       "Tumor_Seq_Allele_1                                                             A   \n",
       "Tumor_Seq_Allele_2                                                             G   \n",
       "dbSNP_RS                                                                     NaN   \n",
       "dbSNP_Val_Status                                                             NaN   \n",
       "Tumor_Sample_Barcode                                TCGA-CG-5723-01A-11D-1600-08   \n",
       "Matched_Norm_Sample_Barcode                         TCGA-CG-5723-11A-02D-1600-08   \n",
       "Match_Norm_Seq_Allele1                                                       NaN   \n",
       "Match_Norm_Seq_Allele2                                                       NaN   \n",
       "Tumor_Validation_Allele1                                                     NaN   \n",
       "Tumor_Validation_Allele2                                                     NaN   \n",
       "Match_Norm_Validation_Allele1                                                NaN   \n",
       "Match_Norm_Validation_Allele2                                                NaN   \n",
       "Verification_Status                                                          NaN   \n",
       "Validation_Status                                                       Untested   \n",
       "Mutation_Status                                                          Somatic   \n",
       "Sequencing_Phase                                                         Phase_I   \n",
       "Sequence_Source                                                              WXS   \n",
       "...                                                                          ...   \n",
       "n_DNA_Matching_Both                                                            0   \n",
       "n_RNA_Matching_Reference                                                      83   \n",
       "n_RNA_Matching_Tumor                                                          36   \n",
       "n_RNA_Matching_Neither                                                        12   \n",
       "n_RNA_Matching_Both                                                            0   \n",
       "tumorDNAFraction                                                       0.1460674   \n",
       "tumorRNAFraction                                                       0.2748092   \n",
       "tumorDNAMultiple                                                       -5.846154   \n",
       "tumorRNAMultiple                                                       -2.638889   \n",
       "tumorDNARatio                                                          0.1710526   \n",
       "tumorRNARatio                                                          0.3789474   \n",
       "FractionOfMeanExpression                                                1.139024   \n",
       "zOfmeanExpression                                                       0.235176   \n",
       "IsSingle                                                                    True   \n",
       "CancerType                                                                  stad   \n",
       "gender                                                                      male   \n",
       "gene                                                                        TP53   \n",
       "start                                                                    7578424   \n",
       "end                                                                      7578424   \n",
       "chr                                                                        chr17   \n",
       "var_allele                                                                     A   \n",
       "chrom                                                                      chr17   \n",
       "pos                                                                      7521788   \n",
       "id                                          7d61ad22-c96f-11e4-936d-4929716f0807   \n",
       "ref                                                                            G   \n",
       "alt                                                                            T   \n",
       "qual                                                                           .   \n",
       "filter                                                                 SR;RP;VUM   \n",
       "normal                                              0|0:0:0:15:0:0:0:2:0:0.0e+00   \n",
       "tumor                                              0|1:0:0:20:3:0:0:14:2:1.3e-01   \n",
       "\n",
       "                                                                        12220598  \n",
       "RNAFile                        lusc\\548da644-a577-4c86-b04e-2ca21fe4c774-TP53...  \n",
       "DNAFile                        lusc\\1ccbfc36-db90-477a-9333-c733b12ff52c-TP53...  \n",
       "Hugo_Symbol                                                                 TP53  \n",
       "Entrez_Gene_Id                                                              7157  \n",
       "Center                                                             broad.mit.edu  \n",
       "NCBI_Build                                                                    37  \n",
       "Chromosome                                                                    17  \n",
       "Start_Position                                                           7579470  \n",
       "End_Position                                                             7579470  \n",
       "Strand                                                                         +  \n",
       "Variant_Classification                                           Frame_Shift_Del  \n",
       "Variant_Type                                                                 DEL  \n",
       "Reference_Allele                                                               C  \n",
       "Tumor_Seq_Allele_1                                                             C  \n",
       "Tumor_Seq_Allele_2                                                             -  \n",
       "dbSNP_RS                                                                     NaN  \n",
       "dbSNP_Val_Status                                                             NaN  \n",
       "Tumor_Sample_Barcode                                TCGA-60-2713-01A-01D-1522-08  \n",
       "Matched_Norm_Sample_Barcode                         TCGA-60-2713-11A-01D-1522-08  \n",
       "Match_Norm_Seq_Allele1                                                       NaN  \n",
       "Match_Norm_Seq_Allele2                                                       NaN  \n",
       "Tumor_Validation_Allele1                                                     NaN  \n",
       "Tumor_Validation_Allele2                                                     NaN  \n",
       "Match_Norm_Validation_Allele1                                                NaN  \n",
       "Match_Norm_Validation_Allele2                                                NaN  \n",
       "Verification_Status                                                          NaN  \n",
       "Validation_Status                                                       Untested  \n",
       "Mutation_Status                                                          Somatic  \n",
       "Sequencing_Phase                                                         Phase_I  \n",
       "Sequence_Source                                                              WXS  \n",
       "...                                                                          ...  \n",
       "n_DNA_Matching_Both                                                            0  \n",
       "n_RNA_Matching_Reference                                                      33  \n",
       "n_RNA_Matching_Tumor                                                           0  \n",
       "n_RNA_Matching_Neither                                                        34  \n",
       "n_RNA_Matching_Both                                                            0  \n",
       "tumorDNAFraction                                                      0.03448276  \n",
       "tumorRNAFraction                                                               0  \n",
       "tumorDNAMultiple                                                             -28  \n",
       "tumorRNAMultiple                                                             -30  \n",
       "tumorDNARatio                                                         0.03571429  \n",
       "tumorRNARatio                                                             0.0001  \n",
       "FractionOfMeanExpression                                               0.3217284  \n",
       "zOfmeanExpression                                                     -0.8527172  \n",
       "IsSingle                                                                    True  \n",
       "CancerType                                                                  lusc  \n",
       "gender                                                                      male  \n",
       "gene                                                                        TP53  \n",
       "start                                                                    7579470  \n",
       "end                                                                      7579470  \n",
       "chr                                                                        chr17  \n",
       "var_allele                                                                     C  \n",
       "chrom                                                                      chr17  \n",
       "pos                                                                      7521767  \n",
       "id                                          7cd88364-ceef-11e4-86db-c482ec69b06c  \n",
       "ref                                                                            T  \n",
       "alt                                                                            G  \n",
       "qual                                                                           .  \n",
       "filter                                                              SR;MN;RP;VUM  \n",
       "normal                                              0|0:0:0:0:5:0:0:0:17:0.0e+00  \n",
       "tumor                                                0|1:0:0:3:1:0:0:0:6:3.0e-01  \n",
       "\n",
       "[72 rows x 8 columns]"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutvars_p53_enh = mutvars[(mutvars.chrom=='chr17') & (mutvars.pos >= 7520200) & (mutvars.pos <= 7521800)]\n",
    "mutvars_p53_enh.transpose()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HGNC ID</th>\n",
       "      <th>Approved Symbol</th>\n",
       "      <th>Approved Name</th>\n",
       "      <th>Status</th>\n",
       "      <th>Previous Symbols</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Accession Numbers</th>\n",
       "      <th>Entrez Gene ID</th>\n",
       "      <th>Ensembl Gene ID</th>\n",
       "      <th>RefSeq IDs</th>\n",
       "      <th>Vega ID</th>\n",
       "      <th>OMIM ID(supplied by OMIM)</th>\n",
       "      <th>UniProt ID(supplied by UniProt)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16964</th>\n",
       "      <td>HGNC:11998</td>\n",
       "      <td>TP53</td>\n",
       "      <td>tumor protein p53</td>\n",
       "      <td>Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>p53, LFS1</td>\n",
       "      <td>17p13.1</td>\n",
       "      <td>AF307851</td>\n",
       "      <td>7157</td>\n",
       "      <td>ENSG00000141510</td>\n",
       "      <td>NM_000546</td>\n",
       "      <td>OTTHUMG00000162125</td>\n",
       "      <td>191170</td>\n",
       "      <td>P04637</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          HGNC ID Approved Symbol      Approved Name    Status  \\\n",
       "16964  HGNC:11998            TP53  tumor protein p53  Approved   \n",
       "\n",
       "      Previous Symbols   Synonyms Chromosome Accession Numbers  \\\n",
       "16964              NaN  p53, LFS1    17p13.1          AF307851   \n",
       "\n",
       "       Entrez Gene ID  Ensembl Gene ID RefSeq IDs             Vega ID  \\\n",
       "16964            7157  ENSG00000141510  NM_000546  OTTHUMG00000162125   \n",
       "\n",
       "      OMIM ID(supplied by OMIM) UniProt ID(supplied by UniProt)  \n",
       "16964                    191170                          P04637  "
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hgnc = pd.read_table('c:/ravip/rstuff/hgnc.tsv')\n",
    "hgnc[hgnc['Approved Symbol']=='TP53']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
